{"cells":[{"cell_type":"code","source":["#First importing the required packages\n","\n","!pip install lightning\n","\n","!pip install torchmetrics\n","\n","!pip install obonet\n","\n","!pip install pyvis\n","\n","!pip install bio"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"iYHhhKiEDGhr","executionInfo":{"status":"ok","timestamp":1711126059747,"user_tz":300,"elapsed":190848,"user":{"displayName":"Harshavardhana Srinivasan","userId":"07445518868972480492"}},"outputId":"73683dab-6503-411f-e6e8-d95c9d06192c"},"id":"iYHhhKiEDGhr","execution_count":1,"outputs":[{"output_type":"stream","name":"stdout","text":["Collecting lightning\n","  Downloading lightning-2.2.1-py3-none-any.whl (2.1 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m2.1/2.1 MB\u001b[0m \u001b[31m17.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hRequirement already satisfied: PyYAML<8.0,>=5.4 in /usr/local/lib/python3.10/dist-packages (from lightning) (6.0.1)\n","Requirement already satisfied: fsspec[http]<2025.0,>=2022.5.0 in /usr/local/lib/python3.10/dist-packages (from lightning) (2023.6.0)\n","Collecting lightning-utilities<2.0,>=0.8.0 (from lightning)\n","  Downloading lightning_utilities-0.11.0-py3-none-any.whl (25 kB)\n","Requirement already satisfied: numpy<3.0,>=1.17.2 in /usr/local/lib/python3.10/dist-packages (from lightning) (1.25.2)\n","Requirement already satisfied: packaging<25.0,>=20.0 in /usr/local/lib/python3.10/dist-packages (from lightning) (24.0)\n","Requirement already satisfied: torch<4.0,>=1.13.0 in /usr/local/lib/python3.10/dist-packages (from lightning) (2.2.1+cu121)\n","Collecting torchmetrics<3.0,>=0.7.0 (from lightning)\n","  Downloading torchmetrics-1.3.2-py3-none-any.whl (841 kB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m841.5/841.5 kB\u001b[0m \u001b[31m13.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hRequirement already satisfied: tqdm<6.0,>=4.57.0 in /usr/local/lib/python3.10/dist-packages (from lightning) (4.66.2)\n","Requirement already satisfied: typing-extensions<6.0,>=4.4.0 in /usr/local/lib/python3.10/dist-packages (from lightning) (4.10.0)\n","Collecting pytorch-lightning (from lightning)\n","  Downloading pytorch_lightning-2.2.1-py3-none-any.whl (801 kB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m801.6/801.6 kB\u001b[0m \u001b[31m16.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hRequirement already satisfied: requests in /usr/local/lib/python3.10/dist-packages (from fsspec[http]<2025.0,>=2022.5.0->lightning) (2.31.0)\n","Requirement already satisfied: aiohttp!=4.0.0a0,!=4.0.0a1 in /usr/local/lib/python3.10/dist-packages (from fsspec[http]<2025.0,>=2022.5.0->lightning) (3.9.3)\n","Requirement already satisfied: setuptools in /usr/local/lib/python3.10/dist-packages (from lightning-utilities<2.0,>=0.8.0->lightning) (67.7.2)\n","Requirement already satisfied: filelock in /usr/local/lib/python3.10/dist-packages (from torch<4.0,>=1.13.0->lightning) (3.13.1)\n","Requirement already satisfied: sympy in /usr/local/lib/python3.10/dist-packages (from torch<4.0,>=1.13.0->lightning) (1.12)\n","Requirement already satisfied: networkx in /usr/local/lib/python3.10/dist-packages (from torch<4.0,>=1.13.0->lightning) (3.2.1)\n","Requirement already satisfied: jinja2 in /usr/local/lib/python3.10/dist-packages (from torch<4.0,>=1.13.0->lightning) (3.1.3)\n","Collecting nvidia-cuda-nvrtc-cu12==12.1.105 (from torch<4.0,>=1.13.0->lightning)\n","  Downloading nvidia_cuda_nvrtc_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (23.7 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m23.7/23.7 MB\u001b[0m \u001b[31m8.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting nvidia-cuda-runtime-cu12==12.1.105 (from torch<4.0,>=1.13.0->lightning)\n","  Downloading nvidia_cuda_runtime_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (823 kB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m823.6/823.6 kB\u001b[0m \u001b[31m14.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting nvidia-cuda-cupti-cu12==12.1.105 (from torch<4.0,>=1.13.0->lightning)\n","  Downloading nvidia_cuda_cupti_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (14.1 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m14.1/14.1 MB\u001b[0m \u001b[31m17.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting nvidia-cudnn-cu12==8.9.2.26 (from torch<4.0,>=1.13.0->lightning)\n","  Downloading nvidia_cudnn_cu12-8.9.2.26-py3-none-manylinux1_x86_64.whl (731.7 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m731.7/731.7 MB\u001b[0m \u001b[31m678.3 kB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting nvidia-cublas-cu12==12.1.3.1 (from torch<4.0,>=1.13.0->lightning)\n","  Downloading nvidia_cublas_cu12-12.1.3.1-py3-none-manylinux1_x86_64.whl (410.6 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m410.6/410.6 MB\u001b[0m \u001b[31m2.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting nvidia-cufft-cu12==11.0.2.54 (from torch<4.0,>=1.13.0->lightning)\n","  Downloading nvidia_cufft_cu12-11.0.2.54-py3-none-manylinux1_x86_64.whl (121.6 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m121.6/121.6 MB\u001b[0m \u001b[31m8.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting nvidia-curand-cu12==10.3.2.106 (from torch<4.0,>=1.13.0->lightning)\n","  Downloading nvidia_curand_cu12-10.3.2.106-py3-none-manylinux1_x86_64.whl (56.5 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m56.5/56.5 MB\u001b[0m \u001b[31m10.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting nvidia-cusolver-cu12==11.4.5.107 (from torch<4.0,>=1.13.0->lightning)\n","  Downloading nvidia_cusolver_cu12-11.4.5.107-py3-none-manylinux1_x86_64.whl (124.2 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m124.2/124.2 MB\u001b[0m \u001b[31m4.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting nvidia-cusparse-cu12==12.1.0.106 (from torch<4.0,>=1.13.0->lightning)\n","  Downloading nvidia_cusparse_cu12-12.1.0.106-py3-none-manylinux1_x86_64.whl (196.0 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m196.0/196.0 MB\u001b[0m \u001b[31m1.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting nvidia-nccl-cu12==2.19.3 (from torch<4.0,>=1.13.0->lightning)\n","  Downloading nvidia_nccl_cu12-2.19.3-py3-none-manylinux1_x86_64.whl (166.0 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m166.0/166.0 MB\u001b[0m \u001b[31m5.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting nvidia-nvtx-cu12==12.1.105 (from torch<4.0,>=1.13.0->lightning)\n","  Downloading nvidia_nvtx_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (99 kB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m99.1/99.1 kB\u001b[0m \u001b[31m9.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hRequirement already satisfied: triton==2.2.0 in /usr/local/lib/python3.10/dist-packages (from torch<4.0,>=1.13.0->lightning) (2.2.0)\n","Collecting nvidia-nvjitlink-cu12 (from nvidia-cusolver-cu12==11.4.5.107->torch<4.0,>=1.13.0->lightning)\n","  Downloading nvidia_nvjitlink_cu12-12.4.99-py3-none-manylinux2014_x86_64.whl (21.1 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m21.1/21.1 MB\u001b[0m \u001b[31m40.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hRequirement already satisfied: aiosignal>=1.1.2 in /usr/local/lib/python3.10/dist-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]<2025.0,>=2022.5.0->lightning) (1.3.1)\n","Requirement already satisfied: attrs>=17.3.0 in /usr/local/lib/python3.10/dist-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]<2025.0,>=2022.5.0->lightning) (23.2.0)\n","Requirement already satisfied: frozenlist>=1.1.1 in /usr/local/lib/python3.10/dist-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]<2025.0,>=2022.5.0->lightning) (1.4.1)\n","Requirement already satisfied: multidict<7.0,>=4.5 in /usr/local/lib/python3.10/dist-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]<2025.0,>=2022.5.0->lightning) (6.0.5)\n","Requirement already satisfied: yarl<2.0,>=1.0 in /usr/local/lib/python3.10/dist-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]<2025.0,>=2022.5.0->lightning) (1.9.4)\n","Requirement already satisfied: async-timeout<5.0,>=4.0 in /usr/local/lib/python3.10/dist-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]<2025.0,>=2022.5.0->lightning) (4.0.3)\n","Requirement already satisfied: MarkupSafe>=2.0 in /usr/local/lib/python3.10/dist-packages (from jinja2->torch<4.0,>=1.13.0->lightning) (2.1.5)\n","Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.10/dist-packages (from requests->fsspec[http]<2025.0,>=2022.5.0->lightning) (3.3.2)\n","Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.10/dist-packages (from requests->fsspec[http]<2025.0,>=2022.5.0->lightning) (3.6)\n","Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.10/dist-packages (from requests->fsspec[http]<2025.0,>=2022.5.0->lightning) (2.0.7)\n","Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.10/dist-packages (from requests->fsspec[http]<2025.0,>=2022.5.0->lightning) (2024.2.2)\n","Requirement already satisfied: mpmath>=0.19 in /usr/local/lib/python3.10/dist-packages (from sympy->torch<4.0,>=1.13.0->lightning) (1.3.0)\n","Installing collected packages: nvidia-nvtx-cu12, nvidia-nvjitlink-cu12, nvidia-nccl-cu12, nvidia-curand-cu12, nvidia-cufft-cu12, nvidia-cuda-runtime-cu12, nvidia-cuda-nvrtc-cu12, nvidia-cuda-cupti-cu12, nvidia-cublas-cu12, lightning-utilities, nvidia-cusparse-cu12, nvidia-cudnn-cu12, nvidia-cusolver-cu12, torchmetrics, pytorch-lightning, lightning\n","Successfully installed lightning-2.2.1 lightning-utilities-0.11.0 nvidia-cublas-cu12-12.1.3.1 nvidia-cuda-cupti-cu12-12.1.105 nvidia-cuda-nvrtc-cu12-12.1.105 nvidia-cuda-runtime-cu12-12.1.105 nvidia-cudnn-cu12-8.9.2.26 nvidia-cufft-cu12-11.0.2.54 nvidia-curand-cu12-10.3.2.106 nvidia-cusolver-cu12-11.4.5.107 nvidia-cusparse-cu12-12.1.0.106 nvidia-nccl-cu12-2.19.3 nvidia-nvjitlink-cu12-12.4.99 nvidia-nvtx-cu12-12.1.105 pytorch-lightning-2.2.1 torchmetrics-1.3.2\n","Requirement already satisfied: torchmetrics in /usr/local/lib/python3.10/dist-packages (1.3.2)\n","Requirement already satisfied: numpy>1.20.0 in /usr/local/lib/python3.10/dist-packages (from torchmetrics) (1.25.2)\n","Requirement already satisfied: packaging>17.1 in /usr/local/lib/python3.10/dist-packages (from torchmetrics) (24.0)\n","Requirement already satisfied: torch>=1.10.0 in /usr/local/lib/python3.10/dist-packages (from torchmetrics) (2.2.1+cu121)\n","Requirement already satisfied: lightning-utilities>=0.8.0 in /usr/local/lib/python3.10/dist-packages (from torchmetrics) (0.11.0)\n","Requirement already satisfied: setuptools in /usr/local/lib/python3.10/dist-packages (from lightning-utilities>=0.8.0->torchmetrics) (67.7.2)\n","Requirement already satisfied: typing-extensions in /usr/local/lib/python3.10/dist-packages (from lightning-utilities>=0.8.0->torchmetrics) (4.10.0)\n","Requirement already satisfied: filelock in /usr/local/lib/python3.10/dist-packages (from torch>=1.10.0->torchmetrics) (3.13.1)\n","Requirement already satisfied: sympy in /usr/local/lib/python3.10/dist-packages (from torch>=1.10.0->torchmetrics) (1.12)\n","Requirement already satisfied: networkx in /usr/local/lib/python3.10/dist-packages (from torch>=1.10.0->torchmetrics) (3.2.1)\n","Requirement already satisfied: jinja2 in /usr/local/lib/python3.10/dist-packages (from torch>=1.10.0->torchmetrics) (3.1.3)\n","Requirement already satisfied: fsspec in /usr/local/lib/python3.10/dist-packages (from torch>=1.10.0->torchmetrics) (2023.6.0)\n","Requirement already satisfied: nvidia-cuda-nvrtc-cu12==12.1.105 in /usr/local/lib/python3.10/dist-packages (from torch>=1.10.0->torchmetrics) (12.1.105)\n","Requirement already satisfied: nvidia-cuda-runtime-cu12==12.1.105 in /usr/local/lib/python3.10/dist-packages (from torch>=1.10.0->torchmetrics) (12.1.105)\n","Requirement already satisfied: nvidia-cuda-cupti-cu12==12.1.105 in /usr/local/lib/python3.10/dist-packages (from torch>=1.10.0->torchmetrics) (12.1.105)\n","Requirement already satisfied: nvidia-cudnn-cu12==8.9.2.26 in /usr/local/lib/python3.10/dist-packages (from torch>=1.10.0->torchmetrics) (8.9.2.26)\n","Requirement already satisfied: nvidia-cublas-cu12==12.1.3.1 in /usr/local/lib/python3.10/dist-packages (from torch>=1.10.0->torchmetrics) (12.1.3.1)\n","Requirement already satisfied: nvidia-cufft-cu12==11.0.2.54 in /usr/local/lib/python3.10/dist-packages (from torch>=1.10.0->torchmetrics) (11.0.2.54)\n","Requirement already satisfied: nvidia-curand-cu12==10.3.2.106 in /usr/local/lib/python3.10/dist-packages (from torch>=1.10.0->torchmetrics) (10.3.2.106)\n","Requirement already satisfied: nvidia-cusolver-cu12==11.4.5.107 in /usr/local/lib/python3.10/dist-packages (from torch>=1.10.0->torchmetrics) (11.4.5.107)\n","Requirement already satisfied: nvidia-cusparse-cu12==12.1.0.106 in /usr/local/lib/python3.10/dist-packages (from torch>=1.10.0->torchmetrics) (12.1.0.106)\n","Requirement already satisfied: nvidia-nccl-cu12==2.19.3 in /usr/local/lib/python3.10/dist-packages (from torch>=1.10.0->torchmetrics) (2.19.3)\n","Requirement already satisfied: nvidia-nvtx-cu12==12.1.105 in /usr/local/lib/python3.10/dist-packages (from torch>=1.10.0->torchmetrics) (12.1.105)\n","Requirement already satisfied: triton==2.2.0 in /usr/local/lib/python3.10/dist-packages (from torch>=1.10.0->torchmetrics) (2.2.0)\n","Requirement already satisfied: nvidia-nvjitlink-cu12 in /usr/local/lib/python3.10/dist-packages (from nvidia-cusolver-cu12==11.4.5.107->torch>=1.10.0->torchmetrics) (12.4.99)\n","Requirement already satisfied: MarkupSafe>=2.0 in /usr/local/lib/python3.10/dist-packages (from jinja2->torch>=1.10.0->torchmetrics) (2.1.5)\n","Requirement already satisfied: mpmath>=0.19 in /usr/local/lib/python3.10/dist-packages (from sympy->torch>=1.10.0->torchmetrics) (1.3.0)\n","Collecting obonet\n","  Downloading obonet-1.0.0-py3-none-any.whl (9.2 kB)\n","Requirement already satisfied: networkx in /usr/local/lib/python3.10/dist-packages (from obonet) (3.2.1)\n","Installing collected packages: obonet\n","Successfully installed obonet-1.0.0\n","Collecting pyvis\n","  Downloading pyvis-0.3.2-py3-none-any.whl (756 kB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m756.0/756.0 kB\u001b[0m \u001b[31m6.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hRequirement already satisfied: ipython>=5.3.0 in /usr/local/lib/python3.10/dist-packages (from pyvis) (7.34.0)\n","Requirement already satisfied: jinja2>=2.9.6 in /usr/local/lib/python3.10/dist-packages (from pyvis) (3.1.3)\n","Requirement already satisfied: jsonpickle>=1.4.1 in /usr/local/lib/python3.10/dist-packages (from pyvis) (3.0.3)\n","Requirement already satisfied: networkx>=1.11 in /usr/local/lib/python3.10/dist-packages (from pyvis) (3.2.1)\n","Requirement already satisfied: setuptools>=18.5 in /usr/local/lib/python3.10/dist-packages (from ipython>=5.3.0->pyvis) (67.7.2)\n","Collecting jedi>=0.16 (from ipython>=5.3.0->pyvis)\n","  Downloading jedi-0.19.1-py2.py3-none-any.whl (1.6 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.6/1.6 MB\u001b[0m \u001b[31m11.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hRequirement already satisfied: decorator in /usr/local/lib/python3.10/dist-packages (from ipython>=5.3.0->pyvis) (4.4.2)\n","Requirement already satisfied: pickleshare in /usr/local/lib/python3.10/dist-packages (from ipython>=5.3.0->pyvis) (0.7.5)\n","Requirement already satisfied: traitlets>=4.2 in /usr/local/lib/python3.10/dist-packages (from ipython>=5.3.0->pyvis) (5.7.1)\n","Requirement already satisfied: prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0 in /usr/local/lib/python3.10/dist-packages (from ipython>=5.3.0->pyvis) (3.0.43)\n","Requirement already satisfied: pygments in /usr/local/lib/python3.10/dist-packages (from ipython>=5.3.0->pyvis) (2.16.1)\n","Requirement already satisfied: backcall in /usr/local/lib/python3.10/dist-packages (from ipython>=5.3.0->pyvis) (0.2.0)\n","Requirement already satisfied: matplotlib-inline in /usr/local/lib/python3.10/dist-packages (from ipython>=5.3.0->pyvis) (0.1.6)\n","Requirement already satisfied: pexpect>4.3 in /usr/local/lib/python3.10/dist-packages (from ipython>=5.3.0->pyvis) (4.9.0)\n","Requirement already satisfied: MarkupSafe>=2.0 in /usr/local/lib/python3.10/dist-packages (from jinja2>=2.9.6->pyvis) (2.1.5)\n","Requirement already satisfied: parso<0.9.0,>=0.8.3 in /usr/local/lib/python3.10/dist-packages (from jedi>=0.16->ipython>=5.3.0->pyvis) (0.8.3)\n","Requirement already satisfied: ptyprocess>=0.5 in /usr/local/lib/python3.10/dist-packages (from pexpect>4.3->ipython>=5.3.0->pyvis) (0.7.0)\n","Requirement already satisfied: wcwidth in /usr/local/lib/python3.10/dist-packages (from prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0->ipython>=5.3.0->pyvis) (0.2.13)\n","Installing collected packages: jedi, pyvis\n","Successfully installed jedi-0.19.1 pyvis-0.3.2\n","Collecting bio\n","  Downloading bio-1.6.2-py3-none-any.whl (278 kB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m278.6/278.6 kB\u001b[0m \u001b[31m4.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hCollecting biopython>=1.80 (from bio)\n","  Downloading biopython-1.83-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.1 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.1/3.1 MB\u001b[0m \u001b[31m13.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hRequirement already satisfied: requests in /usr/local/lib/python3.10/dist-packages (from bio) (2.31.0)\n","Requirement already satisfied: tqdm in /usr/local/lib/python3.10/dist-packages (from bio) (4.66.2)\n","Collecting mygene (from bio)\n","  Downloading mygene-3.2.2-py2.py3-none-any.whl (5.4 kB)\n","Requirement already satisfied: pandas in /usr/local/lib/python3.10/dist-packages (from bio) (1.5.3)\n","Requirement already satisfied: pooch in /usr/local/lib/python3.10/dist-packages (from bio) (1.8.1)\n","Collecting gprofiler-official (from bio)\n","  Downloading gprofiler_official-1.0.0-py3-none-any.whl (9.3 kB)\n","Requirement already satisfied: numpy in /usr/local/lib/python3.10/dist-packages (from biopython>=1.80->bio) (1.25.2)\n","Collecting biothings-client>=0.2.6 (from mygene->bio)\n","  Downloading biothings_client-0.3.1-py2.py3-none-any.whl (29 kB)\n","Requirement already satisfied: python-dateutil>=2.8.1 in /usr/local/lib/python3.10/dist-packages (from pandas->bio) (2.8.2)\n","Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.10/dist-packages (from pandas->bio) (2023.4)\n","Requirement already satisfied: platformdirs>=2.5.0 in /usr/local/lib/python3.10/dist-packages (from pooch->bio) (4.2.0)\n","Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.10/dist-packages (from pooch->bio) (24.0)\n","Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.10/dist-packages (from requests->bio) (3.3.2)\n","Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.10/dist-packages (from requests->bio) (3.6)\n","Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.10/dist-packages (from requests->bio) (2.0.7)\n","Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.10/dist-packages (from requests->bio) (2024.2.2)\n","Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.10/dist-packages (from python-dateutil>=2.8.1->pandas->bio) (1.16.0)\n","Installing collected packages: biopython, gprofiler-official, biothings-client, mygene, bio\n","Successfully installed bio-1.6.2 biopython-1.83 biothings-client-0.3.1 gprofiler-official-1.0.0 mygene-3.2.2\n"]}]},{"cell_type":"code","execution_count":3,"id":"aa7f2959-0de0-4762-bd5b-3be8db82f5f5","metadata":{"scrolled":true,"id":"aa7f2959-0de0-4762-bd5b-3be8db82f5f5","executionInfo":{"status":"ok","timestamp":1711126138638,"user_tz":300,"elapsed":313,"user":{"displayName":"Harshavardhana Srinivasan","userId":"07445518868972480492"}}},"outputs":[],"source":["import numpy as np\n","import pandas as pd\n","import matplotlib.pyplot as plt\n","import seaborn as sns\n","from tqdm import tqdm\n","import os\n","import random\n","\n","# Graphs & Protein sequences\n","import networkx as nx\n","import obonet\n","from pyvis.network import Network\n","from Bio import SeqIO\n","\n","# Deep Learning\n","import torch\n","import torch.nn as nn\n","from torch.utils.data import Dataset, DataLoader\n","from torch.utils.data import random_split\n","from torch.optim.lr_scheduler import ReduceLROnPlateau\n","from torchmetrics.classification import MultilabelF1Score\n","from torchmetrics.classification import MultilabelAccuracy\n","import lightning as L\n","\n","# Embeddings\n","# from transformers import AutoTokenizer, BioGptModel\n","\n","# biogpt_tokenizer = AutoTokenizer.from_pretrained(\"microsoft/biogpt\")\n","# biogpt_model = BioGptModel.from_pretrained(\"microsoft/biogpt\")\n","\n","# Formatting\n","from IPython.core.interactiveshell import InteractiveShell\n","InteractiveShell.ast_node_interactivity = \"all\""]},{"cell_type":"code","source":["from google.colab import drive\n","drive.mount('/content/drive/')"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":316},"id":"VjQv9plfEQiW","executionInfo":{"status":"error","timestamp":1711126227660,"user_tz":300,"elapsed":30580,"user":{"displayName":"Harshavardhana Srinivasan","userId":"07445518868972480492"}},"outputId":"f7b078b1-318b-414d-97d9-d2f3e0094efd"},"id":"VjQv9plfEQiW","execution_count":5,"outputs":[{"output_type":"error","ename":"MessageError","evalue":"Error: credential propagation was unsuccessful","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mMessageError\u001b[0m                              Traceback (most recent call last)","\u001b[0;32m<ipython-input-5-91874b305a32>\u001b[0m in \u001b[0;36m<cell line: 2>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mgoogle\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolab\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mdrive\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mdrive\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmount\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'/content/drive/'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/google/colab/drive.py\u001b[0m in \u001b[0;36mmount\u001b[0;34m(mountpoint, force_remount, timeout_ms, readonly)\u001b[0m\n\u001b[1;32m     98\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mmount\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmountpoint\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mforce_remount\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtimeout_ms\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m120000\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreadonly\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     99\u001b[0m   \u001b[0;34m\"\"\"Mount your Google Drive at the specified mountpoint path.\"\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 100\u001b[0;31m   return _mount(\n\u001b[0m\u001b[1;32m    101\u001b[0m       \u001b[0mmountpoint\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    102\u001b[0m       \u001b[0mforce_remount\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mforce_remount\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/google/colab/drive.py\u001b[0m in \u001b[0;36m_mount\u001b[0;34m(mountpoint, force_remount, timeout_ms, ephemeral, readonly)\u001b[0m\n\u001b[1;32m    131\u001b[0m   )\n\u001b[1;32m    132\u001b[0m   \u001b[0;32mif\u001b[0m \u001b[0mephemeral\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 133\u001b[0;31m     _message.blocking_request(\n\u001b[0m\u001b[1;32m    134\u001b[0m         \u001b[0;34m'request_auth'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrequest\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m{\u001b[0m\u001b[0;34m'authType'\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;34m'dfs_ephemeral'\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtimeout_sec\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    135\u001b[0m     )\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/google/colab/_message.py\u001b[0m in \u001b[0;36mblocking_request\u001b[0;34m(request_type, request, timeout_sec, parent)\u001b[0m\n\u001b[1;32m    174\u001b[0m       \u001b[0mrequest_type\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrequest\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mparent\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mparent\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mexpect_reply\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    175\u001b[0m   )\n\u001b[0;32m--> 176\u001b[0;31m   \u001b[0;32mreturn\u001b[0m \u001b[0mread_reply_from_input\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrequest_id\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtimeout_sec\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/google/colab/_message.py\u001b[0m in \u001b[0;36mread_reply_from_input\u001b[0;34m(message_id, timeout_sec)\u001b[0m\n\u001b[1;32m    101\u001b[0m     ):\n\u001b[1;32m    102\u001b[0m       \u001b[0;32mif\u001b[0m \u001b[0;34m'error'\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mreply\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 103\u001b[0;31m         \u001b[0;32mraise\u001b[0m \u001b[0mMessageError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mreply\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'error'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    104\u001b[0m       \u001b[0;32mreturn\u001b[0m \u001b[0mreply\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'data'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    105\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mMessageError\u001b[0m: Error: credential propagation was unsuccessful"]}]},{"cell_type":"code","execution_count":null,"id":"ad8c89fb-b43d-4329-85aa-2f26ee205346","metadata":{"id":"ad8c89fb-b43d-4329-85aa-2f26ee205346"},"outputs":[],"source":["train_root = \"Train\"\n","test_root = \"Test (Targets)\""]},{"cell_type":"markdown","id":"9d6cb686-ca40-4a21-a964-5acf31eda7a6","metadata":{"id":"9d6cb686-ca40-4a21-a964-5acf31eda7a6"},"source":["# Task Description\n","\n","The objective of our model is to predict the terms (functions) of a protein sequence. One protein sequence can have many functions and can thus be classified into any number of terms. Each term is uniquely identified by a GO Term ID. Thus, our model has to predict all the GO Term IDs for a protein sequence, which means that the task at hand is a **multi-label classification** problem."]},{"cell_type":"markdown","id":"c9adeb53-5f2c-4e84-b494-1d95aeef77a1","metadata":{"id":"c9adeb53-5f2c-4e84-b494-1d95aeef77a1"},"source":["# File Descriptions"]},{"cell_type":"markdown","id":"b98b566b-91aa-474f-9deb-030721d0a9b0","metadata":{"jp-MarkdownHeadingCollapsed":true,"id":"b98b566b-91aa-474f-9deb-030721d0a9b0"},"source":["## Training Data"]},{"cell_type":"markdown","id":"7a5b26e1-65d4-48c1-8321-685fbd3f60d2","metadata":{"id":"7a5b26e1-65d4-48c1-8321-685fbd3f60d2"},"source":["### About the $X$'s"]},{"cell_type":"markdown","id":"fbe9a956-737a-4fa3-a29b-4acabeb2b741","metadata":{"id":"fbe9a956-737a-4fa3-a29b-4acabeb2b741"},"source":["#### 1. `train_sequences.fasta`"]},{"cell_type":"markdown","id":"8c129bf2-0e4b-4d65-b871-c6a6a9e1a696","metadata":{"id":"8c129bf2-0e4b-4d65-b871-c6a6a9e1a696"},"source":["* Contains the protein sequences for the training dataset\n","    * Most protein sequences were extracted from the Swiss-Prot database, but a subset of proteins that are not represented in Swiss-Prot were extracted from the TrEMBL database. More info [here](https://www.uniprot.org/help/uniprotkb_sections).\n","* Header indicates from which database the sequence originate, either Swiss-Prot or TrEMBL, both being parts of UniProtKB.\n","    * For example, `sp|P9WHI7|RECN_MYCT` in the FASTA header indicates the protein...\n","        1. with UniProt ID `P9WHI7`\n","        2. with entry name `RECN_MYCT`\n","        3. taken from Swiss-Prot (`sp`)\n","            * Any sequences taken from TrEMBL will have `tr` in the header instead of sp.\n","    * Detailed format of the fasta header can be found [here](https://www.uniprot.org/help/fasta-headers)\n","* This file contains only sequences for proteins with annotations in the dataset (labeled proteins). To obtain the full set of protein sequences for unlabeled proteins, the Swiss-Prot and TrEMBL databases can be found [here](https://www.uniprot.org/help/downloads)."]},{"cell_type":"markdown","id":"daadaccd-7e02-4232-ae9f-3145d8a9fe55","metadata":{"id":"daadaccd-7e02-4232-ae9f-3145d8a9fe55"},"source":["#### 2. `train_taxonomy.tsv`"]},{"cell_type":"markdown","id":"2e4117c8-12c4-4ad9-9353-273da0383964","metadata":{"id":"2e4117c8-12c4-4ad9-9353-273da0383964"},"source":["* Contains the list of **proteins** and the **species to which they belong**, represented by a \"taxonomic identifier\" (taxon ID) number.\n","    * The first column `EntryID`: the protein UniProt accession ID\n","    * The second column `taxonomyID`: the taxon ID\n","* More information about taxonomies can he found [here](https://www.uniprot.org/help/taxonomic_identifier)."]},{"cell_type":"markdown","id":"e45aa338-9100-4d90-9e72-0f1c2e0ba1ce","metadata":{"id":"e45aa338-9100-4d90-9e72-0f1c2e0ba1ce"},"source":["### About the $y$'s"]},{"cell_type":"markdown","id":"6372a57d-0c93-421f-97b7-7d3525252cbe","metadata":{"id":"6372a57d-0c93-421f-97b7-7d3525252cbe"},"source":["#### 3. `train_terms.tsv`"]},{"cell_type":"markdown","id":"19903228-2eb2-4d75-a270-8e703169ade6","metadata":{"jp-MarkdownHeadingCollapsed":true,"id":"19903228-2eb2-4d75-a270-8e703169ade6"},"source":["* Contains the list of **annotated terms (ground truth)** for the proteins in `train_sequences.fasta`.\n","    * The first column `EntryID`: the protein's UniProt accession ID\n","    * The second column `term`: the GO term ID\n","    * The third column `aspect`: the ontology in which the term appears"]},{"cell_type":"markdown","id":"b4838e61-52de-4c6d-8a14-c80cd335ab30","metadata":{"jp-MarkdownHeadingCollapsed":true,"id":"b4838e61-52de-4c6d-8a14-c80cd335ab30"},"source":["#### 4. `go-basic.obo`"]},{"cell_type":"markdown","id":"42c22804-ace6-44ef-90d1-d0d383ea07ae","metadata":{"id":"42c22804-ace6-44ef-90d1-d0d383ea07ae"},"source":["* The Gene Ontology (GO) data in OBO format, which indicates the **functional properties of proteins**. If a protein is labeled with a term, it means that this protein has this function validated by experimental evidence. A protein can have multiple labels. The absence of a term annotation does not necessarily mean a protein does not have this function, only that this annotation does not exist (yet) in the GO.\n","* GO describes our understanding of the biological domain with respect to the below three aspects, represented as subgraphs. Biologically, each subgraph represents a different aspect of the protein's function: what it does on a molecular level (MF), which biological processes it participates in (BP) and where in the cell it is located (CC).\n","\n","    1. **Molecular Function (MF)**\n","        1. Activities that occur at the molecular level, such as “catalysis” or “transport”\n","        2. generally correspond to activities that can be performed by individual gene products (i.e. a protein or RNA), but some activities are performed by molecular complexes composed of multiple gene products:\n","            1. Examples of broad functional terms: *catalytic activity* and *ransporter activity*\n","            2. Examples of narrower functional terms: *adenylate cyclase activity* or *Toll-like receptor binding*\n","    \n","    2. **Cellular Component (CC)**\n","        1. Locations of gene products, relative to cellular compartments and structures, occupied by a macromolecular machine\n","        2. There are two ways to describe:\n","            1. *Cellular anatomical entities*, in which a gene product carries out a molecular function, including:\n","                1. Cellular structures, such as the plasma membrane and the cytoskeleton\n","                2. Membrane-enclosed cellular compartments, such as the mitochondrion\n","            2. the stable *Macromolecular complexes* of which they are parts\n","    \n","    3. **Biological Process (BP)**\n","        1. The larger processes, or ‘biological programs’ accomplished by multiple molecular activities\n","            1. Examples of broad biological process terms: DNA repair or signal transduction\n","            2. Examples of more specific terms: pyrimidine nucleobase biosynthetic process or glucose transmembrane transport  \n","        \n","* The three GO ontologies are \"*is-a disjoint*\", meaning that no *is-a* relations operate between terms from the different ontologies.\n","    * However, other relationships such as *part-of* and *regulates* do operate between the GO ontologies.\n","    * For example, the molecular function term ‘cyclin-dependent protein kinase activity’ is part of the biological process ‘cell cycle’.\n","* A term may have more than one parent term!!\n","* More about gene ontology [here](https://geneontology.github.io/docs/ontology-documentation/)\n","* This is the 2023-01-01 release of the GO graph\n","    * The nodes in this graph are indexed by the term name:\n","    > ```python\n","      subontology_roots = {'BPO':'GO:0008150',  # Biological Process (BP)\n","                           'CCO':'GO:0005575',  # Cellular Component (CC)\n","                           'MFO':'GO:0003674'}  # Molecular Function (MF)\n","      ```"]},{"cell_type":"markdown","id":"59cf5297-61ff-4348-b8cf-257ddadc24c7","metadata":{"jp-MarkdownHeadingCollapsed":true,"id":"59cf5297-61ff-4348-b8cf-257ddadc24c7"},"source":["#### 5. `IA.txt` - Information Accretion"]},{"cell_type":"markdown","id":"acb615cf-14d7-45c1-81c5-a3a96af0a2de","metadata":{"id":"acb615cf-14d7-45c1-81c5-a3a96af0a2de"},"source":["* Contains the **information accretion (weights) for each GO term**.\n","* These weights are used to compute weighted precision and recall, as described in the Evaluation section.\n","* The values of this file were computed using the following code [repo](https://github.com/claradepaolis/InformationAccretion).\n","    * Information accretion ($ia$), introduced in [Clark and Radivojac, 2013](https://watermark.silverchair.com/bioinformatics_29_13_i53.pdf?token=AQECAHi208BE49Ooan9kkhW_Ercy7Dm3ZL_9Cf3qfKAc485ysgAAA38wggN7BgkqhkiG9w0BBwagggNsMIIDaAIBADCCA2EGCSqGSIb3DQEHATAeBglghkgBZQMEAS4wEQQM_6yBfcYBeK9gGSVMAgEQgIIDMu5V50JH8avL5_BSrU_NcCzyt7-4kGItwHrgn3SubEw_I-Jxi7uzqi-OLZOfba85vy2uD1UkBZokynBNWyxYEuWInuJnDRtbfmwsw7gTmsBiEMwd81gVpfSYeGCERK2J_79hEvsFrdvCPvOUiOmnSrJXYkta_h4TgDREr1IVMwYJcWtLv_Kcj5uwko9BTo226PiWJsV4GkTSdgYNpIqsbbrK28ffYaNIOXG99jnNVu_Z_cLbtzqdU-ZelaiYCzbpFL-dO8Rtwc90pVrxpOrFPcdHirHTkXJaZrUO0fFat-SpGThD5Vnk7If7CJWRHSoNHYaLlv5G1Tm3_vIrHO-tAWbGiU-rYuMs7ye5Dl1GzD0tucx1TveVLWNnj_hgurQVfbFhMjpgl3iuLoVIYi9b_Wv5bSpptksQERMXMTK9a-_l4L4M2w2OpXpECvp-ZLTOml784hHBHelaKfQvZMxwn1kGsEgMLH1b_jiAc7MNWCvwLj4Us3hz1z6hEz0Sv5oPXZsYd9iV-3e1k5fTUmLhPLUXtRoYwKFHtKhAa8X5XnP6b6Y-XNy0CXm-fAuvLLDrxeA2d_uz7ED6RLMrxCjqYEwmS-uiYjbCz05W6_tgd7JQwqwUFAYeDT8-YM0-MQPv5pWOuG4N1PUa54iifeVMFazWdEJDGvEUR5vSNl2RHCc1g9WQ7sCEBqapNAn4jNQDq_yqGttPJYc9EpdPHTfSq5E4tWRwzO_uV0xx5axldiN1DZMlL23hCl1ISSrOWPx60auCvI10fNX0Yldtet29fJ702lDzJ_igwUQ6dkGYWprOtUKl0vmIQ36DRAinj6Fvi7AKmFuVicW4CJPlDlpCj3quai9ApsNX-PIRfJufa7kaVOhppdBKBTh6gEa63KWsd3RQXko44RspOqr_jAVXVxFLtni_4Ferk-U9kWbnkFflFHCD4Wh11GUQaW6ZSYPXow1aebJz4QSik01KdRg86jv0A0PXwO_Ybb1uWln5XaBU2Olj9BwpLP2r5Tkr3nrfZDb-7mo9MwBqmuwRYG4BsDN4IC0OhUmaGsks3m618K2A7CYPdMhKgnX_gbx3xNC4Dpq9), is a measure of how much information is added to an ontology annotation by node $v$ given that its parents $\\mathcal{P}_a(v)$ are already annotated. Specifically,\n","$$\n","ia(v) = \\log_2 \\frac{1}{Pr(v|\\mathcal{P}_a(v))} = \\log_2 \\frac{Pr(\\mathcal{P}_a(v))}{Pr(\\mathcal{P}_a(v)|v) \\cdot Pr(v)}\n","$$\n","* [A great post](https://www.kaggle.com/competitions/cafa-5-protein-function-prediction/discussion/405237) about the intuition behind IA."]},{"cell_type":"markdown","id":"74d9da31-7aff-478c-a05a-62fe9e2e1cfb","metadata":{"jp-MarkdownHeadingCollapsed":true,"id":"74d9da31-7aff-478c-a05a-62fe9e2e1cfb"},"source":["## Testing Data"]},{"cell_type":"markdown","id":"482223ef-dfb5-43de-a931-ee7ee4e3d83a","metadata":{"id":"482223ef-dfb5-43de-a931-ee7ee4e3d83a"},"source":["### About the $X$'s"]},{"cell_type":"markdown","id":"36fc805c-d6ef-40db-8ca2-d6e9cfb74adc","metadata":{"jp-MarkdownHeadingCollapsed":true,"id":"36fc805c-d6ef-40db-8ca2-d6e9cfb74adc"},"source":["#### 1. `testsuperset.fasta`"]},{"cell_type":"markdown","id":"4c0b7225-08cf-4096-857a-e1ec0143886b","metadata":{"jp-MarkdownHeadingCollapsed":true,"id":"4c0b7225-08cf-4096-857a-e1ec0143886b"},"source":["* Contains protein sequences on which the participants are asked to submit predictions\n","* Header contains the protein's UniProt accession ID and the Taxon ID of the species this protein belongs to\n","* Only a small subset of those sequences will accumulate functional annotations and will constitute the test set"]},{"cell_type":"markdown","id":"76ee2dc5-6398-4781-a3df-8e16014999f7","metadata":{"jp-MarkdownHeadingCollapsed":true,"id":"76ee2dc5-6398-4781-a3df-8e16014999f7"},"source":["#### 2. `testsuperset-taxon-list.tsv`"]},{"cell_type":"markdown","id":"2b72af7a-8e85-427a-b216-9743e865aa8c","metadata":{"jp-MarkdownHeadingCollapsed":true,"id":"2b72af7a-8e85-427a-b216-9743e865aa8c"},"source":["* A set of taxon IDs for the proteins in the test superset."]},{"cell_type":"markdown","id":"c645dd71-f793-415c-a1ad-8e9f51230404","metadata":{"jp-MarkdownHeadingCollapsed":true,"id":"c645dd71-f793-415c-a1ad-8e9f51230404"},"source":["# First Look at the data"]},{"cell_type":"code","execution_count":null,"id":"9a0772c8-f442-4f33-9e24-bb94dddc2136","metadata":{"id":"9a0772c8-f442-4f33-9e24-bb94dddc2136"},"outputs":[],"source":["# Read in train data\n","train_sequences_fasta = list(SeqIO.parse(os.path.join(train_root, \"train_sequences.fasta\"), \"fasta\"))\n","train_taxonomy = pd.read_csv(os.path.join(train_root, \"train_taxonomy.tsv\"), sep=\"\\t\")\n","\n","train_terms = pd.read_csv(os.path.join(train_root, \"train_terms.tsv\"), sep=\"\\t\")\n","\n","with open(os.path.join(train_root, \"go-basic.obo\")) as obo_file:\n","    go_graph = obonet.read_obo(obo_file)\n","\n","information_accretion = pd.read_csv(\"IA.txt\", sep=\"\\t\", header=None, names=[\"GO_term\", \"IA\"])\n","\n","# Read in test data\n","test_sequences_fasta = list(SeqIO.parse(os.path.join(test_root, \"testsuperset.fasta\"), \"fasta\"))\n","test_taxonomy = pd.read_csv(os.path.join(test_root, \"testsuperset-taxon-list.tsv\"), sep=\"\\t\", encoding=\"ISO-8859-1\")"]},{"cell_type":"code","execution_count":null,"id":"c967b686-8a56-454a-b34a-224ccac28d87","metadata":{"id":"c967b686-8a56-454a-b34a-224ccac28d87"},"outputs":[],"source":["def plot_dag(graph, term, radius=1):\n","    \"\"\"Create a smaller subgraph with `radius`-hop neighbors given a center node `term`\"\"\"\n","    ng_graph = nx.ego_graph(graph, term, radius=radius)\n","\n","    for node in ng_graph.nodes(data=True):\n","        # concatenate label of the node with its attribute\n","        node[1][\"label\"] = node[0] + \" \" + node[1][\"name\"]\n","\n","    net = Network(\n","        directed=True,\n","        notebook=True,\n","        # cdn_resources=\"in_line\",\n","        height=\"750px\",\n","        width=\"100%\",\n","    )\n","    net.from_nx(ng_graph)\n","\n","    return net.show(\"network.html\")"]},{"cell_type":"code","execution_count":null,"id":"e6492e78-e631-4079-8a62-530c19c1c8d7","metadata":{"id":"e6492e78-e631-4079-8a62-530c19c1c8d7","outputId":"922c047a-6936-4f11-f71e-7a9a78de8e0b"},"outputs":[{"name":"stdout","output_type":"stream","text":["train_sequences_fasta:\n","P20536\n","P20536 sp|P20536|UNG_VACCC Uracil-DNA glycosylase OS=Vaccinia virus (strain Copenhagen) OX=10249 GN=UNG PE=1 SV=1\n","Seq('MNSVTVSHAPYTITYHDDWEPVMSQLVEFYNEVASWLLRDETSPIPDKFFIQLK...FIY')\n","218\n","============================================================\n","train_taxonomy:\n"]},{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>EntryID</th>\n","      <th>taxonomyID</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>Q8IXT2</td>\n","      <td>9606</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>Q04418</td>\n","      <td>559292</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>A8DYA3</td>\n","      <td>7227</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>Q9UUI3</td>\n","      <td>284812</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>Q57ZS4</td>\n","      <td>185431</td>\n","    </tr>\n","  </tbody>\n","</table>\n","</div>"],"text/plain":["  EntryID  taxonomyID\n","0  Q8IXT2        9606\n","1  Q04418      559292\n","2  A8DYA3        7227\n","3  Q9UUI3      284812\n","4  Q57ZS4      185431"]},"execution_count":5,"metadata":{},"output_type":"execute_result"},{"data":{"text/plain":["(142246, 2)"]},"execution_count":5,"metadata":{},"output_type":"execute_result"},{"name":"stdout","output_type":"stream","text":["============================================================\n","train_terms:\n"]},{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>EntryID</th>\n","      <th>term</th>\n","      <th>aspect</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0008152</td>\n","      <td>BPO</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0034655</td>\n","      <td>BPO</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0072523</td>\n","      <td>BPO</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0044270</td>\n","      <td>BPO</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0006753</td>\n","      <td>BPO</td>\n","    </tr>\n","  </tbody>\n","</table>\n","</div>"],"text/plain":["      EntryID        term aspect\n","0  A0A009IHW8  GO:0008152    BPO\n","1  A0A009IHW8  GO:0034655    BPO\n","2  A0A009IHW8  GO:0072523    BPO\n","3  A0A009IHW8  GO:0044270    BPO\n","4  A0A009IHW8  GO:0006753    BPO"]},"execution_count":5,"metadata":{},"output_type":"execute_result"},{"data":{"text/plain":["(5363863, 3)"]},"execution_count":5,"metadata":{},"output_type":"execute_result"},{"name":"stdout","output_type":"stream","text":["============================================================\n","go_graph\n","Gene Ontology: <class 'networkx.classes.multidigraph.MultiDiGraph'>\n","Number of nodes: 43248\n","Number of edges: 84805\n","============================================================\n","information_accretion\n"]},{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>GO_term</th>\n","      <th>IA</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>GO:0000001</td>\n","      <td>0.000000</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>GO:0000002</td>\n","      <td>3.103836</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>GO:0000003</td>\n","      <td>3.439404</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>GO:0000011</td>\n","      <td>0.056584</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>GO:0000012</td>\n","      <td>6.400377</td>\n","    </tr>\n","  </tbody>\n","</table>\n","</div>"],"text/plain":["      GO_term        IA\n","0  GO:0000001  0.000000\n","1  GO:0000002  3.103836\n","2  GO:0000003  3.439404\n","3  GO:0000011  0.056584\n","4  GO:0000012  6.400377"]},"execution_count":5,"metadata":{},"output_type":"execute_result"},{"data":{"text/plain":["(43248, 2)"]},"execution_count":5,"metadata":{},"output_type":"execute_result"},{"name":"stdout","output_type":"stream","text":["============================================================\n","test_sequences_fasta\n","Q9CQV8\n","Q9CQV8\t10090\n","Seq('MTMDKSELVQKAKLAEQAERYDDMAAAMKAVTEQGHELSNEERNLLSVAYKNVV...GEN')\n","246\n","============================================================\n","test_taxonomy:\n"]},{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>ID</th>\n","      <th>Species</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>9606</td>\n","      <td>homo sapiens[All Names]</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>10090</td>\n","      <td>mus musculus[All Names]</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>10116</td>\n","      <td>Rattus norvegicus</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>3702</td>\n","      <td>Arabidopsis thaliana[All Names]</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>83333</td>\n","      <td>Escherichia coli K-12[all names]</td>\n","    </tr>\n","  </tbody>\n","</table>\n","</div>"],"text/plain":["      ID                           Species\n","0   9606           homo sapiens[All Names]\n","1  10090           mus musculus[All Names]\n","2  10116                 Rattus norvegicus\n","3   3702   Arabidopsis thaliana[All Names]\n","4  83333  Escherichia coli K-12[all names]"]},"execution_count":5,"metadata":{},"output_type":"execute_result"},{"data":{"text/plain":["(90, 2)"]},"execution_count":5,"metadata":{},"output_type":"execute_result"}],"source":["# train_sequences_fasta\n","print(\"train_sequences_fasta:\")\n","for seq_record in train_sequences_fasta:\n","    print(seq_record.id)\n","    print(seq_record.description)\n","    print(repr(seq_record.seq))\n","    print(len(seq_record))\n","    break\n","print(\"=\" * 60)\n","\n","# train_taxonomy\n","print(\"train_taxonomy:\")\n","train_taxonomy.head()\n","train_taxonomy.shape\n","print(\"=\" * 60)\n","\n","# train_terms\n","print(\"train_terms:\")\n","train_terms.head()\n","train_terms.shape\n","print(\"=\" * 60)\n","\n","# go_graph\n","print(\"go_graph\")\n","print(f\"Gene Ontology: {type(go_graph)}\")\n","print(f\"Number of nodes: {go_graph.number_of_nodes()}\")\n","print(f\"Number of edges: {go_graph.number_of_edges()}\")\n","print(\"=\" * 60)\n","\n","# information_accretion\n","print(\"information_accretion\")\n","information_accretion.head()\n","information_accretion.shape\n","print(\"=\" * 60)\n","\n","# test_sequences_fasta\n","print(\"test_sequences_fasta\")\n","for seq_record in test_sequences_fasta:\n","    print(seq_record.id)\n","    print(seq_record.description)\n","    print(repr(seq_record.seq))\n","    print(len(seq_record))\n","    break\n","print(\"=\" * 60)\n","\n","# test_taxonomy\n","print(\"test_taxonomy:\")\n","test_taxonomy.head()\n","test_taxonomy.shape"]},{"cell_type":"markdown","id":"45227519-4a94-4130-8d09-8ddac027aae6","metadata":{"id":"45227519-4a94-4130-8d09-8ddac027aae6"},"source":["# Explore the GO graph and NetworkX (library) functionalities"]},{"cell_type":"markdown","id":"36414e73-398d-47dd-b6a3-5a897de96616","metadata":{"id":"36414e73-398d-47dd-b6a3-5a897de96616"},"source":["#### What is in each node (GO term)?"]},{"cell_type":"markdown","id":"40fa490f-621f-4a5c-9aa4-c9864eda7993","metadata":{"jp-MarkdownHeadingCollapsed":true,"id":"40fa490f-621f-4a5c-9aa4-c9864eda7993"},"source":["##### Essential elements"]},{"cell_type":"markdown","id":"541632bb-c93e-4e45-b43e-de6c6e4167a3","metadata":{"id":"541632bb-c93e-4e45-b43e-de6c6e4167a3"},"source":["* `term`: The unique identifier of the GO term\n","* `name`: a human-readable name for the GO term\n","* `namespace`: the aspect - which of the three sub-ontologies (CC, BP or MF) the term belongs to.\n","* `def`: A textual description of what the term represents, plus reference(s) to the source of the information.\n","* `is_a`: Relationships to other terms in the ontology.\n","    * All terms (other than the root terms representing each aspect) have an *is a* sub-class relationship to another term\n","    * For example, *glucose transmembrane transport (GO:1904659)* is a *monosaccharide transport (GO:0015749)*.\n","    * The Gene Ontology employs a number of other relations; the relations documentation page describes the relations used in the ontology."]},{"cell_type":"markdown","id":"3d5fa628-0217-4ea9-a649-55f71decf114","metadata":{"jp-MarkdownHeadingCollapsed":true,"id":"3d5fa628-0217-4ea9-a649-55f71decf114"},"source":["##### Optional elements"]},{"cell_type":"markdown","id":"9e54fb12-ada6-48ec-8fa6-439c33054cdb","metadata":{"id":"9e54fb12-ada6-48ec-8fa6-439c33054cdb"},"source":["* `alt_id`: Secondary IDs come about when two or more terms are identical in meaning, and are merged into a single term. All terms IDs are preserved so that no information (for example, annotations to the merged IDs) is lost.\n","* `comment`: Any extra information about the term and its usage.\n","* `synonym`: Alternative words or phrases closely related in meaning to the term name, with indication of the relationship between the name and synonym given by the synonym scope. The scopes for GO synonyms are:\n","    * __Exact__: an exact equivalent; interchangeable with the term name; for e.g. ornithine cycle is an exact synonym of urea cycle\n","    * __Broad__: the synonym is broader than the term name; for e.g. cell division is a broad synonym of cytokinesis\n","    * __Narrow__: the synonym is narrower or more precise than the term name; for e.g. pyrimidine-dimer repair by photolyase is a narrow synonym of photoreactive repair\n","    * __Related__: the terms are related in some imprecise way; for e.g. cytochrome bc1 complex is a related synonym of ubiquinol-cytochrome-c reductase activity virulence is a related synonym of pathogenesis.\n","    * Custom synonym types are also used in the ontology. For example, a number of synonyms are designated as systematic synonyms; synonyms of this type are exact synonyms of the term name.\n","* `xref`: Database cross-references, or dbxrefs, refer to identical or very similar objects in other databases.\n","    * For instance, the molecular function term retinal isomerase activity (GO:0004744) is cross-referenced with RHEA:24124\n","    * Another example, the biological process term ulfate assimilation (GO:0000103) has the InterPro cross-reference Sulphate adenylyltransferase (IPR002650).\n","* `subset`: Indicates that the term belongs to a designated subset of terms, e.g. one of the [GO subsets](https://geneontology.org/docs/go-subset-guide/) (also known as GO slims). It offers an overall sense of the key biological functions that are vital to an organism.\n","    * More about GO subsets [here](https://geneontology.org/docs/go-subset-guide/)\n","\n","\n","More information about elements in each node can be found [here](https://geneontology.org/docs/GO-term-elements)"]},{"cell_type":"markdown","id":"e1f3b7c0-0678-469e-b1cf-771a3858ba90","metadata":{"jp-MarkdownHeadingCollapsed":true,"id":"e1f3b7c0-0678-469e-b1cf-771a3858ba90"},"source":["#### Exploration"]},{"cell_type":"code","execution_count":null,"id":"24978aac-7cb5-490c-8148-325ae57a5b68","metadata":{"id":"24978aac-7cb5-490c-8148-325ae57a5b68","outputId":"c4b71b57-eef8-4833-c9f8-d286ede97971"},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>EntryID</th>\n","      <th>term</th>\n","      <th>aspect</th>\n","      <th>GO_name</th>\n","      <th>GO_namespace</th>\n","      <th>GO_definition</th>\n","      <th>GO_is_a</th>\n","      <th>GO_alt_id</th>\n","      <th>GO_comment</th>\n","      <th>GO_xref</th>\n","      <th>GO_subset</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0008152</td>\n","      <td>BPO</td>\n","      <td>metabolic process</td>\n","      <td>biological_process</td>\n","      <td>\"The chemical reactions and pathways, includin...</td>\n","      <td>[GO:0008150]</td>\n","      <td>[GO:0044236, GO:0044710]</td>\n","      <td>Note that metabolic processes do not include s...</td>\n","      <td>[Wikipedia:Metabolism]</td>\n","      <td>[gocheck_do_not_manually_annotate, goslim_chem...</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0034655</td>\n","      <td>BPO</td>\n","      <td>nucleobase-containing compound catabolic process</td>\n","      <td>biological_process</td>\n","      <td>\"The chemical reactions and pathways resulting...</td>\n","      <td>[GO:0006139, GO:0019439, GO:0044270, GO:004670...</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>[goslim_chembl]</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0072523</td>\n","      <td>BPO</td>\n","      <td>purine-containing compound catabolic process</td>\n","      <td>biological_process</td>\n","      <td>\"The chemical reactions and pathways resulting...</td>\n","      <td>[GO:0019439, GO:0044270, GO:0046700, GO:007252...</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0044270</td>\n","      <td>BPO</td>\n","      <td>cellular nitrogen compound catabolic process</td>\n","      <td>biological_process</td>\n","      <td>\"The chemical reactions and pathways resulting...</td>\n","      <td>[GO:0034641, GO:0044248]</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0006753</td>\n","      <td>BPO</td>\n","      <td>nucleoside phosphate metabolic process</td>\n","      <td>biological_process</td>\n","      <td>\"The chemical reactions and pathways involving...</td>\n","      <td>[GO:0006796, GO:0019637, GO:0055086]</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","    </tr>\n","    <tr>\n","      <th>5</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:1901292</td>\n","      <td>BPO</td>\n","      <td>nucleoside phosphate catabolic process</td>\n","      <td>biological_process</td>\n","      <td>\"The chemical reactions and pathways resulting...</td>\n","      <td>[GO:0006753, GO:0034655, GO:0046434]</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","    </tr>\n","    <tr>\n","      <th>6</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0044237</td>\n","      <td>BPO</td>\n","      <td>cellular metabolic process</td>\n","      <td>biological_process</td>\n","      <td>\"The chemical reactions and pathways by which ...</td>\n","      <td>[GO:0008152, GO:0009987]</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>[gocheck_do_not_annotate]</td>\n","    </tr>\n","    <tr>\n","      <th>7</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:1901360</td>\n","      <td>BPO</td>\n","      <td>organic cyclic compound metabolic process</td>\n","      <td>biological_process</td>\n","      <td>\"The chemical reactions and pathways involving...</td>\n","      <td>[GO:0071704]</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","    </tr>\n","    <tr>\n","      <th>8</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0008150</td>\n","      <td>BPO</td>\n","      <td>biological_process</td>\n","      <td>biological_process</td>\n","      <td>\"A biological process represents a specific ob...</td>\n","      <td>NaN</td>\n","      <td>[GO:0000004, GO:0007582, GO:0044699]</td>\n","      <td>Note that, in addition to forming the root of ...</td>\n","      <td>[Wikipedia:Biological_process]</td>\n","      <td>[goslim_candida, goslim_chembl, goslim_metagen...</td>\n","    </tr>\n","    <tr>\n","      <th>9</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:1901564</td>\n","      <td>BPO</td>\n","      <td>organonitrogen compound metabolic process</td>\n","      <td>biological_process</td>\n","      <td>\"The chemical reactions and pathways involving...</td>\n","      <td>[GO:0006807, GO:0071704]</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","    </tr>\n","    <tr>\n","      <th>10</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:1901565</td>\n","      <td>BPO</td>\n","      <td>organonitrogen compound catabolic process</td>\n","      <td>biological_process</td>\n","      <td>\"The chemical reactions and pathways resulting...</td>\n","      <td>[GO:1901564, GO:1901575]</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","    </tr>\n","    <tr>\n","      <th>11</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0009117</td>\n","      <td>BPO</td>\n","      <td>nucleotide metabolic process</td>\n","      <td>biological_process</td>\n","      <td>\"The chemical reactions and pathways involving...</td>\n","      <td>[GO:0006753]</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>[goslim_pir]</td>\n","    </tr>\n","    <tr>\n","      <th>12</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0006139</td>\n","      <td>BPO</td>\n","      <td>nucleobase-containing compound metabolic process</td>\n","      <td>biological_process</td>\n","      <td>\"Any cellular metabolic process involving nucl...</td>\n","      <td>[GO:0006725, GO:0034641, GO:0044238, GO:004648...</td>\n","      <td>[GO:0055134]</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>[goslim_pir, goslim_plant]</td>\n","    </tr>\n","    <tr>\n","      <th>13</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0044281</td>\n","      <td>BPO</td>\n","      <td>small molecule metabolic process</td>\n","      <td>biological_process</td>\n","      <td>\"The chemical reactions and pathways involving...</td>\n","      <td>[GO:0008152]</td>\n","      <td>NaN</td>\n","      <td>Small molecules in GO include monosaccharides ...</td>\n","      <td>NaN</td>\n","      <td>[goslim_chembl, goslim_flybase_ribbon, goslim_...</td>\n","    </tr>\n","    <tr>\n","      <th>14</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0046496</td>\n","      <td>BPO</td>\n","      <td>nicotinamide nucleotide metabolic process</td>\n","      <td>biological_process</td>\n","      <td>\"The chemical reactions and pathways involving...</td>\n","      <td>[GO:0019362]</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","    </tr>\n","    <tr>\n","      <th>15</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0019362</td>\n","      <td>BPO</td>\n","      <td>pyridine nucleotide metabolic process</td>\n","      <td>biological_process</td>\n","      <td>\"The chemical reactions and pathways involving...</td>\n","      <td>[GO:0009117, GO:0072524]</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","    </tr>\n","    <tr>\n","      <th>16</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0046483</td>\n","      <td>BPO</td>\n","      <td>heterocycle metabolic process</td>\n","      <td>biological_process</td>\n","      <td>\"The chemical reactions and pathways involving...</td>\n","      <td>[GO:0044237]</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>[goslim_pir]</td>\n","    </tr>\n","    <tr>\n","      <th>17</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0055086</td>\n","      <td>BPO</td>\n","      <td>nucleobase-containing small molecule metabolic...</td>\n","      <td>biological_process</td>\n","      <td>\"The cellular chemical reactions and pathways ...</td>\n","      <td>[GO:0006139, GO:0044281]</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>[goslim_drosophila, goslim_generic, goslim_pom...</td>\n","    </tr>\n","    <tr>\n","      <th>18</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0044248</td>\n","      <td>BPO</td>\n","      <td>cellular catabolic process</td>\n","      <td>biological_process</td>\n","      <td>\"The chemical reactions and pathways resulting...</td>\n","      <td>[GO:0009056, GO:0044237]</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","    </tr>\n","    <tr>\n","      <th>19</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0019439</td>\n","      <td>BPO</td>\n","      <td>aromatic compound catabolic process</td>\n","      <td>biological_process</td>\n","      <td>\"The chemical reactions and pathways resulting...</td>\n","      <td>[GO:0006725, GO:0044248]</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","    </tr>\n","    <tr>\n","      <th>20</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0019637</td>\n","      <td>BPO</td>\n","      <td>organophosphate metabolic process</td>\n","      <td>biological_process</td>\n","      <td>\"The chemical reactions and pathways involving...</td>\n","      <td>[GO:0006793, GO:0071704]</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","    </tr>\n","    <tr>\n","      <th>21</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0006807</td>\n","      <td>BPO</td>\n","      <td>nitrogen compound metabolic process</td>\n","      <td>biological_process</td>\n","      <td>\"The chemical reactions and pathways involving...</td>\n","      <td>[GO:0008152]</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>[goslim_metagenomics, goslim_pir]</td>\n","    </tr>\n","    <tr>\n","      <th>22</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0019677</td>\n","      <td>BPO</td>\n","      <td>NAD catabolic process</td>\n","      <td>biological_process</td>\n","      <td>\"The chemical reactions and pathways resulting...</td>\n","      <td>[GO:0006195, GO:0019364, GO:0019674]</td>\n","      <td>[GO:0006737]</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","    </tr>\n","    <tr>\n","      <th>23</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:1901361</td>\n","      <td>BPO</td>\n","      <td>organic cyclic compound catabolic process</td>\n","      <td>biological_process</td>\n","      <td>\"The chemical reactions and pathways resulting...</td>\n","      <td>[GO:1901360, GO:1901575]</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","    </tr>\n","    <tr>\n","      <th>24</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0006163</td>\n","      <td>BPO</td>\n","      <td>purine nucleotide metabolic process</td>\n","      <td>biological_process</td>\n","      <td>\"The chemical reactions and pathways involving...</td>\n","      <td>[GO:0009117, GO:0072521]</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","    </tr>\n","    <tr>\n","      <th>25</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0046700</td>\n","      <td>BPO</td>\n","      <td>heterocycle catabolic process</td>\n","      <td>biological_process</td>\n","      <td>\"The chemical reactions and pathways resulting...</td>\n","      <td>[GO:0044248, GO:0046483]</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","    </tr>\n","    <tr>\n","      <th>26</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0009987</td>\n","      <td>BPO</td>\n","      <td>cellular process</td>\n","      <td>biological_process</td>\n","      <td>\"Any process that is carried out at the cellul...</td>\n","      <td>[GO:0008150]</td>\n","      <td>[GO:0008151, GO:0044763, GO:0050875]</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>[gocheck_do_not_annotate, goslim_plant]</td>\n","    </tr>\n","    <tr>\n","      <th>27</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0006725</td>\n","      <td>BPO</td>\n","      <td>cellular aromatic compound metabolic process</td>\n","      <td>biological_process</td>\n","      <td>\"The chemical reactions and pathways involving...</td>\n","      <td>[GO:0044237]</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>[goslim_pir]</td>\n","    </tr>\n","    <tr>\n","      <th>28</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0006796</td>\n","      <td>BPO</td>\n","      <td>phosphate-containing compound metabolic process</td>\n","      <td>biological_process</td>\n","      <td>\"The chemical reactions and pathways involving...</td>\n","      <td>[GO:0006793]</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","    </tr>\n","    <tr>\n","      <th>29</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0034641</td>\n","      <td>BPO</td>\n","      <td>cellular nitrogen compound metabolic process</td>\n","      <td>biological_process</td>\n","      <td>\"The chemical reactions and pathways involving...</td>\n","      <td>[GO:0006807, GO:0044237]</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>[goslim_chembl]</td>\n","    </tr>\n","    <tr>\n","      <th>30</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0072521</td>\n","      <td>BPO</td>\n","      <td>purine-containing compound metabolic process</td>\n","      <td>biological_process</td>\n","      <td>\"The chemical reactions and pathways involving...</td>\n","      <td>[GO:0006725, GO:0034641, GO:0046483, GO:190136...</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","    </tr>\n","    <tr>\n","      <th>31</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0071704</td>\n","      <td>BPO</td>\n","      <td>organic substance metabolic process</td>\n","      <td>biological_process</td>\n","      <td>\"The chemical reactions and pathways involving...</td>\n","      <td>[GO:0008152]</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","    </tr>\n","    <tr>\n","      <th>32</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0019364</td>\n","      <td>BPO</td>\n","      <td>pyridine nucleotide catabolic process</td>\n","      <td>biological_process</td>\n","      <td>\"The chemical reactions and pathways resulting...</td>\n","      <td>[GO:0009166, GO:0019362, GO:0072526]</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","    </tr>\n","    <tr>\n","      <th>33</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:1901575</td>\n","      <td>BPO</td>\n","      <td>organic substance catabolic process</td>\n","      <td>biological_process</td>\n","      <td>\"The chemical reactions and pathways resulting...</td>\n","      <td>[GO:0009056, GO:0071704]</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","    </tr>\n","    <tr>\n","      <th>34</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0072526</td>\n","      <td>BPO</td>\n","      <td>pyridine-containing compound catabolic process</td>\n","      <td>biological_process</td>\n","      <td>\"The chemical reactions and pathways resulting...</td>\n","      <td>[GO:0044270, GO:0046700, GO:0072524, GO:190136...</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","    </tr>\n","    <tr>\n","      <th>35</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0046434</td>\n","      <td>BPO</td>\n","      <td>organophosphate catabolic process</td>\n","      <td>biological_process</td>\n","      <td>\"The chemical reactions and pathways resulting...</td>\n","      <td>[GO:0019637, GO:1901575]</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","    </tr>\n","    <tr>\n","      <th>36</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0009166</td>\n","      <td>BPO</td>\n","      <td>nucleotide catabolic process</td>\n","      <td>biological_process</td>\n","      <td>\"The chemical reactions and pathways resulting...</td>\n","      <td>[GO:0009117, GO:1901292]</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","    </tr>\n","    <tr>\n","      <th>37</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0072524</td>\n","      <td>BPO</td>\n","      <td>pyridine-containing compound metabolic process</td>\n","      <td>biological_process</td>\n","      <td>\"The chemical reactions and pathways involving...</td>\n","      <td>[GO:0034641, GO:0046483, GO:1901360, GO:1901564]</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","    </tr>\n","    <tr>\n","      <th>38</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0006195</td>\n","      <td>BPO</td>\n","      <td>purine nucleotide catabolic process</td>\n","      <td>biological_process</td>\n","      <td>\"The chemical reactions and pathways resulting...</td>\n","      <td>[GO:0006163, GO:0009166, GO:0072523]</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","    </tr>\n","    <tr>\n","      <th>39</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0009056</td>\n","      <td>BPO</td>\n","      <td>catabolic process</td>\n","      <td>biological_process</td>\n","      <td>\"The chemical reactions and pathways resulting...</td>\n","      <td>[GO:0008152]</td>\n","      <td>[GO:0044243, GO:0044712]</td>\n","      <td>NaN</td>\n","      <td>[Wikipedia:Catabolism]</td>\n","      <td>[goslim_agr, goslim_chembl, goslim_plant]</td>\n","    </tr>\n","    <tr>\n","      <th>40</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0044238</td>\n","      <td>BPO</td>\n","      <td>primary metabolic process</td>\n","      <td>biological_process</td>\n","      <td>\"The chemical reactions and pathways involving...</td>\n","      <td>[GO:0008152]</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>[goslim_pir]</td>\n","    </tr>\n","    <tr>\n","      <th>41</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0006793</td>\n","      <td>BPO</td>\n","      <td>phosphorus metabolic process</td>\n","      <td>biological_process</td>\n","      <td>\"The chemical reactions and pathways involving...</td>\n","      <td>[GO:0044237]</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>[goslim_pir]</td>\n","    </tr>\n","    <tr>\n","      <th>42</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0019674</td>\n","      <td>BPO</td>\n","      <td>NAD metabolic process</td>\n","      <td>biological_process</td>\n","      <td>\"The chemical reactions and pathways involving...</td>\n","      <td>[GO:0006163, GO:0046496]</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","    </tr>\n","    <tr>\n","      <th>43</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0003674</td>\n","      <td>MFO</td>\n","      <td>molecular_function</td>\n","      <td>molecular_function</td>\n","      <td>\"A molecular process that can be carried out b...</td>\n","      <td>NaN</td>\n","      <td>[GO:0005554]</td>\n","      <td>Note that, in addition to forming the root of ...</td>\n","      <td>NaN</td>\n","      <td>[goslim_candida, goslim_chembl, goslim_metagen...</td>\n","    </tr>\n","    <tr>\n","      <th>44</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0003953</td>\n","      <td>MFO</td>\n","      <td>NAD+ nucleosidase activity</td>\n","      <td>molecular_function</td>\n","      <td>\"Catalysis of the reaction: NAD+ + H2O = nicot...</td>\n","      <td>[GO:0016799]</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>[Reactome:R-HSA-8870346 \"BST1 hydrolyzes NAD+ ...</td>\n","      <td>NaN</td>\n","    </tr>\n","    <tr>\n","      <th>45</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0016787</td>\n","      <td>MFO</td>\n","      <td>hydrolase activity</td>\n","      <td>molecular_function</td>\n","      <td>\"Catalysis of the hydrolysis of various bonds,...</td>\n","      <td>[GO:0003824]</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>[EC:3.-.-.-, Reactome:R-HSA-1236938 \"Partial p...</td>\n","      <td>[goslim_candida, goslim_chembl, goslim_drosoph...</td>\n","    </tr>\n","    <tr>\n","      <th>46</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0016799</td>\n","      <td>MFO</td>\n","      <td>hydrolase activity, hydrolyzing N-glycosyl com...</td>\n","      <td>molecular_function</td>\n","      <td>\"Catalysis of the hydrolysis of any N-glycosyl...</td>\n","      <td>[GO:0016798]</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>[EC:3.2.2.-]</td>\n","      <td>NaN</td>\n","    </tr>\n","    <tr>\n","      <th>47</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0016798</td>\n","      <td>MFO</td>\n","      <td>hydrolase activity, acting on glycosyl bonds</td>\n","      <td>molecular_function</td>\n","      <td>\"Catalysis of the hydrolysis of any glycosyl b...</td>\n","      <td>[GO:0016787]</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>[EC:3.2.-.-, Reactome:R-HSA-1793176 \"DS is cle...</td>\n","      <td>[goslim_chembl, goslim_drosophila, goslim_yeast]</td>\n","    </tr>\n","    <tr>\n","      <th>48</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0003824</td>\n","      <td>MFO</td>\n","      <td>catalytic activity</td>\n","      <td>molecular_function</td>\n","      <td>\"Catalysis of a biochemical reaction at physio...</td>\n","      <td>[GO:0003674]</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>[Wikipedia:Enzyme]</td>\n","      <td>[goslim_agr, goslim_chembl, goslim_flybase_rib...</td>\n","    </tr>\n","  </tbody>\n","</table>\n","</div>"],"text/plain":["       EntryID        term aspect  \\\n","0   A0A009IHW8  GO:0008152    BPO   \n","1   A0A009IHW8  GO:0034655    BPO   \n","2   A0A009IHW8  GO:0072523    BPO   \n","3   A0A009IHW8  GO:0044270    BPO   \n","4   A0A009IHW8  GO:0006753    BPO   \n","5   A0A009IHW8  GO:1901292    BPO   \n","6   A0A009IHW8  GO:0044237    BPO   \n","7   A0A009IHW8  GO:1901360    BPO   \n","8   A0A009IHW8  GO:0008150    BPO   \n","9   A0A009IHW8  GO:1901564    BPO   \n","10  A0A009IHW8  GO:1901565    BPO   \n","11  A0A009IHW8  GO:0009117    BPO   \n","12  A0A009IHW8  GO:0006139    BPO   \n","13  A0A009IHW8  GO:0044281    BPO   \n","14  A0A009IHW8  GO:0046496    BPO   \n","15  A0A009IHW8  GO:0019362    BPO   \n","16  A0A009IHW8  GO:0046483    BPO   \n","17  A0A009IHW8  GO:0055086    BPO   \n","18  A0A009IHW8  GO:0044248    BPO   \n","19  A0A009IHW8  GO:0019439    BPO   \n","20  A0A009IHW8  GO:0019637    BPO   \n","21  A0A009IHW8  GO:0006807    BPO   \n","22  A0A009IHW8  GO:0019677    BPO   \n","23  A0A009IHW8  GO:1901361    BPO   \n","24  A0A009IHW8  GO:0006163    BPO   \n","25  A0A009IHW8  GO:0046700    BPO   \n","26  A0A009IHW8  GO:0009987    BPO   \n","27  A0A009IHW8  GO:0006725    BPO   \n","28  A0A009IHW8  GO:0006796    BPO   \n","29  A0A009IHW8  GO:0034641    BPO   \n","30  A0A009IHW8  GO:0072521    BPO   \n","31  A0A009IHW8  GO:0071704    BPO   \n","32  A0A009IHW8  GO:0019364    BPO   \n","33  A0A009IHW8  GO:1901575    BPO   \n","34  A0A009IHW8  GO:0072526    BPO   \n","35  A0A009IHW8  GO:0046434    BPO   \n","36  A0A009IHW8  GO:0009166    BPO   \n","37  A0A009IHW8  GO:0072524    BPO   \n","38  A0A009IHW8  GO:0006195    BPO   \n","39  A0A009IHW8  GO:0009056    BPO   \n","40  A0A009IHW8  GO:0044238    BPO   \n","41  A0A009IHW8  GO:0006793    BPO   \n","42  A0A009IHW8  GO:0019674    BPO   \n","43  A0A009IHW8  GO:0003674    MFO   \n","44  A0A009IHW8  GO:0003953    MFO   \n","45  A0A009IHW8  GO:0016787    MFO   \n","46  A0A009IHW8  GO:0016799    MFO   \n","47  A0A009IHW8  GO:0016798    MFO   \n","48  A0A009IHW8  GO:0003824    MFO   \n","\n","                                              GO_name        GO_namespace  \\\n","0                                   metabolic process  biological_process   \n","1    nucleobase-containing compound catabolic process  biological_process   \n","2        purine-containing compound catabolic process  biological_process   \n","3        cellular nitrogen compound catabolic process  biological_process   \n","4              nucleoside phosphate metabolic process  biological_process   \n","5              nucleoside phosphate catabolic process  biological_process   \n","6                          cellular metabolic process  biological_process   \n","7           organic cyclic compound metabolic process  biological_process   \n","8                                  biological_process  biological_process   \n","9           organonitrogen compound metabolic process  biological_process   \n","10          organonitrogen compound catabolic process  biological_process   \n","11                       nucleotide metabolic process  biological_process   \n","12   nucleobase-containing compound metabolic process  biological_process   \n","13                   small molecule metabolic process  biological_process   \n","14          nicotinamide nucleotide metabolic process  biological_process   \n","15              pyridine nucleotide metabolic process  biological_process   \n","16                      heterocycle metabolic process  biological_process   \n","17  nucleobase-containing small molecule metabolic...  biological_process   \n","18                         cellular catabolic process  biological_process   \n","19                aromatic compound catabolic process  biological_process   \n","20                  organophosphate metabolic process  biological_process   \n","21                nitrogen compound metabolic process  biological_process   \n","22                              NAD catabolic process  biological_process   \n","23          organic cyclic compound catabolic process  biological_process   \n","24                purine nucleotide metabolic process  biological_process   \n","25                      heterocycle catabolic process  biological_process   \n","26                                   cellular process  biological_process   \n","27       cellular aromatic compound metabolic process  biological_process   \n","28    phosphate-containing compound metabolic process  biological_process   \n","29       cellular nitrogen compound metabolic process  biological_process   \n","30       purine-containing compound metabolic process  biological_process   \n","31                organic substance metabolic process  biological_process   \n","32              pyridine nucleotide catabolic process  biological_process   \n","33                organic substance catabolic process  biological_process   \n","34     pyridine-containing compound catabolic process  biological_process   \n","35                  organophosphate catabolic process  biological_process   \n","36                       nucleotide catabolic process  biological_process   \n","37     pyridine-containing compound metabolic process  biological_process   \n","38                purine nucleotide catabolic process  biological_process   \n","39                                  catabolic process  biological_process   \n","40                          primary metabolic process  biological_process   \n","41                       phosphorus metabolic process  biological_process   \n","42                              NAD metabolic process  biological_process   \n","43                                 molecular_function  molecular_function   \n","44                         NAD+ nucleosidase activity  molecular_function   \n","45                                 hydrolase activity  molecular_function   \n","46  hydrolase activity, hydrolyzing N-glycosyl com...  molecular_function   \n","47       hydrolase activity, acting on glycosyl bonds  molecular_function   \n","48                                 catalytic activity  molecular_function   \n","\n","                                        GO_definition  \\\n","0   \"The chemical reactions and pathways, includin...   \n","1   \"The chemical reactions and pathways resulting...   \n","2   \"The chemical reactions and pathways resulting...   \n","3   \"The chemical reactions and pathways resulting...   \n","4   \"The chemical reactions and pathways involving...   \n","5   \"The chemical reactions and pathways resulting...   \n","6   \"The chemical reactions and pathways by which ...   \n","7   \"The chemical reactions and pathways involving...   \n","8   \"A biological process represents a specific ob...   \n","9   \"The chemical reactions and pathways involving...   \n","10  \"The chemical reactions and pathways resulting...   \n","11  \"The chemical reactions and pathways involving...   \n","12  \"Any cellular metabolic process involving nucl...   \n","13  \"The chemical reactions and pathways involving...   \n","14  \"The chemical reactions and pathways involving...   \n","15  \"The chemical reactions and pathways involving...   \n","16  \"The chemical reactions and pathways involving...   \n","17  \"The cellular chemical reactions and pathways ...   \n","18  \"The chemical reactions and pathways resulting...   \n","19  \"The chemical reactions and pathways resulting...   \n","20  \"The chemical reactions and pathways involving...   \n","21  \"The chemical reactions and pathways involving...   \n","22  \"The chemical reactions and pathways resulting...   \n","23  \"The chemical reactions and pathways resulting...   \n","24  \"The chemical reactions and pathways involving...   \n","25  \"The chemical reactions and pathways resulting...   \n","26  \"Any process that is carried out at the cellul...   \n","27  \"The chemical reactions and pathways involving...   \n","28  \"The chemical reactions and pathways involving...   \n","29  \"The chemical reactions and pathways involving...   \n","30  \"The chemical reactions and pathways involving...   \n","31  \"The chemical reactions and pathways involving...   \n","32  \"The chemical reactions and pathways resulting...   \n","33  \"The chemical reactions and pathways resulting...   \n","34  \"The chemical reactions and pathways resulting...   \n","35  \"The chemical reactions and pathways resulting...   \n","36  \"The chemical reactions and pathways resulting...   \n","37  \"The chemical reactions and pathways involving...   \n","38  \"The chemical reactions and pathways resulting...   \n","39  \"The chemical reactions and pathways resulting...   \n","40  \"The chemical reactions and pathways involving...   \n","41  \"The chemical reactions and pathways involving...   \n","42  \"The chemical reactions and pathways involving...   \n","43  \"A molecular process that can be carried out b...   \n","44  \"Catalysis of the reaction: NAD+ + H2O = nicot...   \n","45  \"Catalysis of the hydrolysis of various bonds,...   \n","46  \"Catalysis of the hydrolysis of any N-glycosyl...   \n","47  \"Catalysis of the hydrolysis of any glycosyl b...   \n","48  \"Catalysis of a biochemical reaction at physio...   \n","\n","                                              GO_is_a  \\\n","0                                        [GO:0008150]   \n","1   [GO:0006139, GO:0019439, GO:0044270, GO:004670...   \n","2   [GO:0019439, GO:0044270, GO:0046700, GO:007252...   \n","3                            [GO:0034641, GO:0044248]   \n","4                [GO:0006796, GO:0019637, GO:0055086]   \n","5                [GO:0006753, GO:0034655, GO:0046434]   \n","6                            [GO:0008152, GO:0009987]   \n","7                                        [GO:0071704]   \n","8                                                 NaN   \n","9                            [GO:0006807, GO:0071704]   \n","10                           [GO:1901564, GO:1901575]   \n","11                                       [GO:0006753]   \n","12  [GO:0006725, GO:0034641, GO:0044238, GO:004648...   \n","13                                       [GO:0008152]   \n","14                                       [GO:0019362]   \n","15                           [GO:0009117, GO:0072524]   \n","16                                       [GO:0044237]   \n","17                           [GO:0006139, GO:0044281]   \n","18                           [GO:0009056, GO:0044237]   \n","19                           [GO:0006725, GO:0044248]   \n","20                           [GO:0006793, GO:0071704]   \n","21                                       [GO:0008152]   \n","22               [GO:0006195, GO:0019364, GO:0019674]   \n","23                           [GO:1901360, GO:1901575]   \n","24                           [GO:0009117, GO:0072521]   \n","25                           [GO:0044248, GO:0046483]   \n","26                                       [GO:0008150]   \n","27                                       [GO:0044237]   \n","28                                       [GO:0006793]   \n","29                           [GO:0006807, GO:0044237]   \n","30  [GO:0006725, GO:0034641, GO:0046483, GO:190136...   \n","31                                       [GO:0008152]   \n","32               [GO:0009166, GO:0019362, GO:0072526]   \n","33                           [GO:0009056, GO:0071704]   \n","34  [GO:0044270, GO:0046700, GO:0072524, GO:190136...   \n","35                           [GO:0019637, GO:1901575]   \n","36                           [GO:0009117, GO:1901292]   \n","37   [GO:0034641, GO:0046483, GO:1901360, GO:1901564]   \n","38               [GO:0006163, GO:0009166, GO:0072523]   \n","39                                       [GO:0008152]   \n","40                                       [GO:0008152]   \n","41                                       [GO:0044237]   \n","42                           [GO:0006163, GO:0046496]   \n","43                                                NaN   \n","44                                       [GO:0016799]   \n","45                                       [GO:0003824]   \n","46                                       [GO:0016798]   \n","47                                       [GO:0016787]   \n","48                                       [GO:0003674]   \n","\n","                               GO_alt_id  \\\n","0               [GO:0044236, GO:0044710]   \n","1                                    NaN   \n","2                                    NaN   \n","3                                    NaN   \n","4                                    NaN   \n","5                                    NaN   \n","6                                    NaN   \n","7                                    NaN   \n","8   [GO:0000004, GO:0007582, GO:0044699]   \n","9                                    NaN   \n","10                                   NaN   \n","11                                   NaN   \n","12                          [GO:0055134]   \n","13                                   NaN   \n","14                                   NaN   \n","15                                   NaN   \n","16                                   NaN   \n","17                                   NaN   \n","18                                   NaN   \n","19                                   NaN   \n","20                                   NaN   \n","21                                   NaN   \n","22                          [GO:0006737]   \n","23                                   NaN   \n","24                                   NaN   \n","25                                   NaN   \n","26  [GO:0008151, GO:0044763, GO:0050875]   \n","27                                   NaN   \n","28                                   NaN   \n","29                                   NaN   \n","30                                   NaN   \n","31                                   NaN   \n","32                                   NaN   \n","33                                   NaN   \n","34                                   NaN   \n","35                                   NaN   \n","36                                   NaN   \n","37                                   NaN   \n","38                                   NaN   \n","39              [GO:0044243, GO:0044712]   \n","40                                   NaN   \n","41                                   NaN   \n","42                                   NaN   \n","43                          [GO:0005554]   \n","44                                   NaN   \n","45                                   NaN   \n","46                                   NaN   \n","47                                   NaN   \n","48                                   NaN   \n","\n","                                           GO_comment  \\\n","0   Note that metabolic processes do not include s...   \n","1                                                 NaN   \n","2                                                 NaN   \n","3                                                 NaN   \n","4                                                 NaN   \n","5                                                 NaN   \n","6                                                 NaN   \n","7                                                 NaN   \n","8   Note that, in addition to forming the root of ...   \n","9                                                 NaN   \n","10                                                NaN   \n","11                                                NaN   \n","12                                                NaN   \n","13  Small molecules in GO include monosaccharides ...   \n","14                                                NaN   \n","15                                                NaN   \n","16                                                NaN   \n","17                                                NaN   \n","18                                                NaN   \n","19                                                NaN   \n","20                                                NaN   \n","21                                                NaN   \n","22                                                NaN   \n","23                                                NaN   \n","24                                                NaN   \n","25                                                NaN   \n","26                                                NaN   \n","27                                                NaN   \n","28                                                NaN   \n","29                                                NaN   \n","30                                                NaN   \n","31                                                NaN   \n","32                                                NaN   \n","33                                                NaN   \n","34                                                NaN   \n","35                                                NaN   \n","36                                                NaN   \n","37                                                NaN   \n","38                                                NaN   \n","39                                                NaN   \n","40                                                NaN   \n","41                                                NaN   \n","42                                                NaN   \n","43  Note that, in addition to forming the root of ...   \n","44                                                NaN   \n","45                                                NaN   \n","46                                                NaN   \n","47                                                NaN   \n","48                                                NaN   \n","\n","                                              GO_xref  \\\n","0                              [Wikipedia:Metabolism]   \n","1                                                 NaN   \n","2                                                 NaN   \n","3                                                 NaN   \n","4                                                 NaN   \n","5                                                 NaN   \n","6                                                 NaN   \n","7                                                 NaN   \n","8                      [Wikipedia:Biological_process]   \n","9                                                 NaN   \n","10                                                NaN   \n","11                                                NaN   \n","12                                                NaN   \n","13                                                NaN   \n","14                                                NaN   \n","15                                                NaN   \n","16                                                NaN   \n","17                                                NaN   \n","18                                                NaN   \n","19                                                NaN   \n","20                                                NaN   \n","21                                                NaN   \n","22                                                NaN   \n","23                                                NaN   \n","24                                                NaN   \n","25                                                NaN   \n","26                                                NaN   \n","27                                                NaN   \n","28                                                NaN   \n","29                                                NaN   \n","30                                                NaN   \n","31                                                NaN   \n","32                                                NaN   \n","33                                                NaN   \n","34                                                NaN   \n","35                                                NaN   \n","36                                                NaN   \n","37                                                NaN   \n","38                                                NaN   \n","39                             [Wikipedia:Catabolism]   \n","40                                                NaN   \n","41                                                NaN   \n","42                                                NaN   \n","43                                                NaN   \n","44  [Reactome:R-HSA-8870346 \"BST1 hydrolyzes NAD+ ...   \n","45  [EC:3.-.-.-, Reactome:R-HSA-1236938 \"Partial p...   \n","46                                       [EC:3.2.2.-]   \n","47  [EC:3.2.-.-, Reactome:R-HSA-1793176 \"DS is cle...   \n","48                                 [Wikipedia:Enzyme]   \n","\n","                                            GO_subset  \n","0   [gocheck_do_not_manually_annotate, goslim_chem...  \n","1                                     [goslim_chembl]  \n","2                                                 NaN  \n","3                                                 NaN  \n","4                                                 NaN  \n","5                                                 NaN  \n","6                           [gocheck_do_not_annotate]  \n","7                                                 NaN  \n","8   [goslim_candida, goslim_chembl, goslim_metagen...  \n","9                                                 NaN  \n","10                                                NaN  \n","11                                       [goslim_pir]  \n","12                         [goslim_pir, goslim_plant]  \n","13  [goslim_chembl, goslim_flybase_ribbon, goslim_...  \n","14                                                NaN  \n","15                                                NaN  \n","16                                       [goslim_pir]  \n","17  [goslim_drosophila, goslim_generic, goslim_pom...  \n","18                                                NaN  \n","19                                                NaN  \n","20                                                NaN  \n","21                  [goslim_metagenomics, goslim_pir]  \n","22                                                NaN  \n","23                                                NaN  \n","24                                                NaN  \n","25                                                NaN  \n","26            [gocheck_do_not_annotate, goslim_plant]  \n","27                                       [goslim_pir]  \n","28                                                NaN  \n","29                                    [goslim_chembl]  \n","30                                                NaN  \n","31                                                NaN  \n","32                                                NaN  \n","33                                                NaN  \n","34                                                NaN  \n","35                                                NaN  \n","36                                                NaN  \n","37                                                NaN  \n","38                                                NaN  \n","39          [goslim_agr, goslim_chembl, goslim_plant]  \n","40                                       [goslim_pir]  \n","41                                       [goslim_pir]  \n","42                                                NaN  \n","43  [goslim_candida, goslim_chembl, goslim_metagen...  \n","44                                                NaN  \n","45  [goslim_candida, goslim_chembl, goslim_drosoph...  \n","46                                                NaN  \n","47   [goslim_chembl, goslim_drosophila, goslim_yeast]  \n","48  [goslim_agr, goslim_chembl, goslim_flybase_rib...  "]},"execution_count":6,"metadata":{},"output_type":"execute_result"}],"source":["def process_node(node):\n","\n","    # Essential elements\n","    name = node.get(\"name\", np.nan)\n","    namespace = node.get(\"namespace\", np.nan)\n","    definition = node.get(\"def\", np.nan)\n","    is_a = node.get(\"is_a\", np.nan)\n","\n","    # Optional elements\n","    alt_id = node.get(\"alt_id\", np.nan)\n","    comment = node.get(\"comment\", np.nan)\n","    subset = node.get(\"subset\", np.nan)\n","    xref = node.get(\"xref\", np.nan)\n","\n","    node_row = pd.DataFrame({\n","        \"GO_name\": name,\n","        \"GO_namespace\": namespace,\n","        \"GO_definition\": definition,\n","        \"GO_is_a\": [is_a],\n","\n","        \"GO_alt_id\": [alt_id],\n","        \"GO_comment\": comment,\n","        \"GO_xref\": [xref],\n","        \"GO_subset\": [subset],\n","    })\n","\n","    return node_row\n","\n","\n","# An example\n","entry_id = \"A0A009IHW8\"\n","train_terms_subset = train_terms.query(\"EntryID == @entry_id\").reset_index(drop=True).copy()\n","node_info = []\n","for row in train_terms_subset.itertuples(index=False):\n","    node_row = process_node(go_graph.nodes[row.term])\n","    node_info.append(node_row)\n","\n","node_info = pd.concat(node_info, ignore_index=True)\n","pd.concat([train_terms_subset, node_info], axis=1)"]},{"cell_type":"code","execution_count":null,"id":"b4f21f1f-509f-4153-890d-362daf87d482","metadata":{"id":"b4f21f1f-509f-4153-890d-362daf87d482","outputId":"39341e19-e0f3-4cdc-ee6a-34203ad59f54"},"outputs":[{"data":{"text/plain":["{'name': 'nucleobase-containing compound catabolic process',\n"," 'namespace': 'biological_process',\n"," 'def': '\"The chemical reactions and pathways resulting in the breakdown of nucleobases, nucleosides, nucleotides and nucleic acids.\" [GOC:mah]',\n"," 'subset': ['goslim_chembl'],\n"," 'synonym': ['\"nucleobase, nucleoside, nucleotide and nucleic acid breakdown\" EXACT []',\n","  '\"nucleobase, nucleoside, nucleotide and nucleic acid catabolic process\" RELATED [GOC:dph, GOC:tb]',\n","  '\"nucleobase, nucleoside, nucleotide and nucleic acid catabolism\" EXACT []',\n","  '\"nucleobase, nucleoside, nucleotide and nucleic acid degradation\" EXACT []'],\n"," 'is_a': ['GO:0006139',\n","  'GO:0019439',\n","  'GO:0044270',\n","  'GO:0046700',\n","  'GO:1901361']}"]},"metadata":{},"output_type":"display_data"}],"source":["# Visualize a subgraph (with the center being a given term)\n","term = \"GO:0034655\"\n","display(go_graph.nodes[term])\n","# display(go_graph.nodes[\"GO:0048308\"])\n","# display(go_graph.nodes[\"GO:0006996\"])\n","# display(go_graph.nodes[\"GO:0016043\"])\n","# display(go_graph.nodes[\"GO:0071840\"])\n","# display(go_graph.nodes[\"GO:0009987\"])\n","\n","# plot_dag(go_graph, term, radius=5)"]},{"cell_type":"code","execution_count":null,"id":"bd4c9759-7949-4d1b-a75d-11c6748db056","metadata":{"id":"bd4c9759-7949-4d1b-a75d-11c6748db056","outputId":"c159b2b5-8f39-4838-b8de-a5139366acbe"},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>node</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>GO:0000001</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>GO:0000002</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>GO:0000003</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>GO:0000006</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>GO:0000007</td>\n","    </tr>\n","    <tr>\n","      <th>...</th>\n","      <td>...</td>\n","    </tr>\n","    <tr>\n","      <th>43243</th>\n","      <td>GO:2001313</td>\n","    </tr>\n","    <tr>\n","      <th>43244</th>\n","      <td>GO:2001314</td>\n","    </tr>\n","    <tr>\n","      <th>43245</th>\n","      <td>GO:2001315</td>\n","    </tr>\n","    <tr>\n","      <th>43246</th>\n","      <td>GO:2001316</td>\n","    </tr>\n","    <tr>\n","      <th>43247</th>\n","      <td>GO:2001317</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>43248 rows × 1 columns</p>\n","</div>"],"text/plain":["             node\n","0      GO:0000001\n","1      GO:0000002\n","2      GO:0000003\n","3      GO:0000006\n","4      GO:0000007\n","...           ...\n","43243  GO:2001313\n","43244  GO:2001314\n","43245  GO:2001315\n","43246  GO:2001316\n","43247  GO:2001317\n","\n","[43248 rows x 1 columns]"]},"execution_count":8,"metadata":{},"output_type":"execute_result"}],"source":["# Get all nodes\n","pd.DataFrame(go_graph.nodes, columns=[\"node\"])"]},{"cell_type":"code","execution_count":null,"id":"d0b7ce5b-d4c0-47bb-ad68-22968b4c67fd","metadata":{"id":"d0b7ce5b-d4c0-47bb-ad68-22968b4c67fd","outputId":"886194a6-4f6e-4762-a2da-44815922cb7f"},"outputs":[{"data":{"text/plain":["[('GO:0000015', {'GO:1902494': {'is_a': {}}, 'GO:0005829': {'part_of': {}}}),\n"," ('GO:0000016', {'GO:0004553': {'is_a': {}}}),\n"," ('GO:0000017', {'GO:0042946': {'is_a': {}}}),\n"," ('GO:0000018', {'GO:0051052': {'is_a': {}}, 'GO:0006310': {'regulates': {}}}),\n"," ('GO:0000019', {'GO:0000018': {'is_a': {}}, 'GO:0006312': {'regulates': {}}}),\n"," ('GO:0000022',\n","  {'GO:0051231': {'is_a': {}},\n","   'GO:1903047': {'is_a': {}},\n","   'GO:0000070': {'part_of': {}},\n","   'GO:0007052': {'part_of': {}}}),\n"," ('GO:0000023', {'GO:0005984': {'is_a': {}}}),\n"," ('GO:0000024', {'GO:0000023': {'is_a': {}}, 'GO:0046351': {'is_a': {}}})]"]},"execution_count":9,"metadata":{},"output_type":"execute_result"},{"data":{"text/plain":["['GO:0006139', 'GO:0019439', 'GO:0044270', 'GO:0046700', 'GO:1901361']"]},"execution_count":9,"metadata":{},"output_type":"execute_result"},{"data":{"text/plain":["<networkx.classes.multidigraph.MultiDiGraph at 0x2c1bc3700>"]},"execution_count":9,"metadata":{},"output_type":"execute_result"}],"source":["# Get neighbors, neighboring edges (iterator)\n","list(go_graph.adjacency())[10:18] # node, adjacency\n","\n","# Get neighbors of a term\n","list(go_graph.neighbors(term))\n","\n","# Get a subgraph with n-hop neighbors\n","nx.ego_graph(go_graph, term, radius=3)"]},{"cell_type":"code","execution_count":null,"id":"f749077a-8f77-48b1-a4e2-1dcaa791f85f","metadata":{"id":"f749077a-8f77-48b1-a4e2-1dcaa791f85f","outputId":"cb6d7cd6-a36f-4392-c01e-f66c3d816970"},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>node</th>\n","      <th>out_degrees</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>GO:0000001</td>\n","      <td>2</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>GO:0000002</td>\n","      <td>2</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>GO:0000003</td>\n","      <td>8</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>GO:0000006</td>\n","      <td>1</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>GO:0000007</td>\n","      <td>1</td>\n","    </tr>\n","    <tr>\n","      <th>...</th>\n","      <td>...</td>\n","      <td>...</td>\n","    </tr>\n","    <tr>\n","      <th>43243</th>\n","      <td>GO:2001313</td>\n","      <td>5</td>\n","    </tr>\n","    <tr>\n","      <th>43244</th>\n","      <td>GO:2001314</td>\n","      <td>3</td>\n","    </tr>\n","    <tr>\n","      <th>43245</th>\n","      <td>GO:2001315</td>\n","      <td>3</td>\n","    </tr>\n","    <tr>\n","      <th>43246</th>\n","      <td>GO:2001316</td>\n","      <td>6</td>\n","    </tr>\n","    <tr>\n","      <th>43247</th>\n","      <td>GO:2001317</td>\n","      <td>9</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>43248 rows × 2 columns</p>\n","</div>"],"text/plain":["             node  out_degrees\n","0      GO:0000001            2\n","1      GO:0000002            2\n","2      GO:0000003            8\n","3      GO:0000006            1\n","4      GO:0000007            1\n","...           ...          ...\n","43243  GO:2001313            5\n","43244  GO:2001314            3\n","43245  GO:2001315            3\n","43246  GO:2001316            6\n","43247  GO:2001317            9\n","\n","[43248 rows x 2 columns]"]},"execution_count":10,"metadata":{},"output_type":"execute_result"}],"source":["# Get out-degrees (num of parents)\n","pd.DataFrame(go_graph.degree, columns=[\"node\", \"out_degrees\"]) # number of parents"]},{"cell_type":"code","execution_count":null,"id":"00f2ba5d-5195-4ab1-ba3d-bc7a7f29a77e","metadata":{"id":"00f2ba5d-5195-4ab1-ba3d-bc7a7f29a77e","outputId":"17c48a27-8627-4e1c-d62a-5b0e9b6e2ed2"},"outputs":[{"data":{"text/plain":["edge\n","is_a                    69350\n","part_of                  6851\n","regulates                3157\n","negatively_regulates     2729\n","positively_regulates     2718\n","Name: count, dtype: int64"]},"execution_count":11,"metadata":{},"output_type":"execute_result"}],"source":["# Types of edges\n","edges = pd.DataFrame(go_graph.edges, columns=[\"child\", \"parent\", \"edge\"])\n","edges.edge.value_counts(dropna=False)"]},{"cell_type":"code","execution_count":null,"id":"162f1423-637d-4343-9350-66eaa7ee93df","metadata":{"id":"162f1423-637d-4343-9350-66eaa7ee93df","outputId":"8416bbfb-0bad-4f57-ecee-a1ec95bc73a2"},"outputs":[{"name":"stdout","output_type":"stream","text":["5\n","None\n"]}],"source":["# Get shortest path between nodes\n","shortest_path_lengths = dict(nx.all_pairs_shortest_path_length(go_graph))\n","# shortest_path_lengths[\"GO:0000001\"][\"GO:0009987\"] # shortest path between child: 1 and parent: 9987\n","# shortest_path_lengths[\"GO:0000001\"][\"GO:0000002\"] # KeyError if no path exists between them\n","\n","def get_shortest_path(child, parent):\n","    \"\"\"Return the number of hops to get from `child` to `parent`\"\"\"\n","    try:\n","        return shortest_path_lengths[child][parent]  # shortest path between child and parent\n","    except KeyError:\n","        return None                                  # KeyError if no path exists between them\n","    except Exception as e:\n","        raise e\n","\n","print(get_shortest_path(\"GO:0000001\", \"GO:0009987\"))\n","print(get_shortest_path(\"GO:0000001\", \"GO:0000002\"))"]},{"cell_type":"code","execution_count":null,"id":"8c09c9d5-03ad-46f1-a043-00ce7b142b39","metadata":{"id":"8c09c9d5-03ad-46f1-a043-00ce7b142b39","outputId":"566f3fda-e98c-4ea3-c77a-a2bf83b96d76"},"outputs":[{"name":"stdout","output_type":"stream","text":["BP_subgraph: #nodes = 27,941, #edges = 64,536\n","CC_subgraph: #nodes =  4,042, #edges =  6,495\n","MF_subgraph: #nodes = 11,262, #edges = 13,714\n"]}],"source":["# Get subontology graphs\n","subontology_roots = {\n","    'BPO':'GO:0008150',  # Biological Process (BP)\n","    'CCO':'GO:0005575',  # Cellular Component (CC)\n","    'MFO':'GO:0003674',  # Molecular Function (MF)\n","}\n","\n","BP_subgraph = nx.induced_subgraph(go_graph, nx.ancestors(go_graph, source=subontology_roots[\"BPO\"]))\n","CC_subgraph = nx.induced_subgraph(go_graph, nx.ancestors(go_graph, source=subontology_roots[\"CCO\"]))\n","MF_subgraph = nx.induced_subgraph(go_graph, nx.ancestors(go_graph, source=subontology_roots[\"MFO\"]))\n","\n","print(\"BP_subgraph: #nodes = {:,}, #edges = {:,}\".format(BP_subgraph.number_of_nodes(), BP_subgraph.number_of_edges()))\n","print(\"CC_subgraph: #nodes = {:>6,}, #edges = {:>6,}\".format(CC_subgraph.number_of_nodes(), CC_subgraph.number_of_edges()))\n","print(\"MF_subgraph: #nodes = {:,}, #edges = {:,}\".format(MF_subgraph.number_of_nodes(), MF_subgraph.number_of_edges()))\n","\n","# pd.DataFrame(BP_subgraph.edges, columns=[\"child\", \"parent\", \"edge\"])\n","# pd.DataFrame(CC_subgraph.edges, columns=[\"child\", \"parent\", \"edge\"])\n","# pd.DataFrame(MF_subgraph.edges, columns=[\"child\", \"parent\", \"edge\"])"]},{"cell_type":"code","execution_count":null,"id":"b3376c1c-ab1f-40b4-872b-cd04b23b022e","metadata":{"id":"b3376c1c-ab1f-40b4-872b-cd04b23b022e","outputId":"657bdb6c-486e-404a-b1ee-35d4a73a18d6"},"outputs":[{"data":{"text/plain":["3"]},"execution_count":14,"metadata":{},"output_type":"execute_result"}],"source":["nx.number_connected_components(go_graph.to_undirected())"]},{"cell_type":"code","execution_count":null,"id":"d11fd83c-45d0-4e28-bc5c-7de22f22d072","metadata":{"id":"d11fd83c-45d0-4e28-bc5c-7de22f22d072","outputId":"37f19732-0d05-4c19-f54a-1e1963f81279"},"outputs":[{"name":"stdout","output_type":"stream","text":["(('GO:0000001', 'GO:0000001'), 'GO:0000001')\n","(('GO:0000001', 'GO:0006996'), 'GO:0000001')\n","(('GO:0000001', 'GO:0007005'), 'GO:0000001')\n","(('GO:0000001', 'GO:0008150'), 'GO:0000001')\n","(('GO:0000001', 'GO:0009987'), 'GO:0000001')\n","(('GO:0000001', 'GO:0016043'), 'GO:0000001')\n","(('GO:0000001', 'GO:0048308'), 'GO:0000001')\n","(('GO:0000001', 'GO:0048311'), 'GO:0000001')\n","(('GO:0000001', 'GO:0051179'), 'GO:0000001')\n","(('GO:0000001', 'GO:0051640'), 'GO:0000001')\n","(('GO:0000001', 'GO:0051646'), 'GO:0000001')\n"]}],"source":["for i, pairs in enumerate(nx.all_pairs_lowest_common_ancestor(go_graph)):\n","    print(pairs)\n","    if i == 10:\n","        break"]},{"cell_type":"code","execution_count":null,"id":"1d89cacf-9bff-4433-9625-eb15dd5a4bd8","metadata":{"id":"1d89cacf-9bff-4433-9625-eb15dd5a4bd8","outputId":"30e903d0-f8f8-434f-fb08-359ff493d81a"},"outputs":[{"data":{"text/plain":["True"]},"execution_count":16,"metadata":{},"output_type":"execute_result"}],"source":["# Check if is DAG\n","nx.is_directed_acyclic_graph(go_graph)"]},{"cell_type":"code","execution_count":null,"id":"c53e21ce-66c7-40ab-a550-64714ca3c835","metadata":{"id":"c53e21ce-66c7-40ab-a550-64714ca3c835","outputId":"9a34b07b-e7c5-42b8-ce3d-6a813ed2ea35"},"outputs":[{"data":{"text/plain":["18"]},"execution_count":17,"metadata":{},"output_type":"execute_result"},{"data":{"text/plain":["14"]},"execution_count":17,"metadata":{},"output_type":"execute_result"},{"data":{"text/plain":["12"]},"execution_count":17,"metadata":{},"output_type":"execute_result"}],"source":["# Topological sort of the graphs\n","# Kahn’s algorithm: It first finds a list of “start nodes” which have no incoming edges\n","BP_top_sorts = list(nx.topological_generations(BP_subgraph))\n","CC_top_sorts = list(nx.topological_generations(CC_subgraph))\n","MF_top_sorts = list(nx.topological_generations(MF_subgraph))\n","\n","len(BP_top_sorts)\n","len(CC_top_sorts)\n","len(MF_top_sorts)"]},{"cell_type":"code","execution_count":null,"id":"0d0b8d6f-1423-4d98-b5e3-e0670b03be64","metadata":{"id":"0d0b8d6f-1423-4d98-b5e3-e0670b03be64","outputId":"12480471-f6da-4dac-9d89-88247143cafa"},"outputs":[{"data":{"text/plain":["[('GO:0110165', 0.009827271255809652),\n"," ('GO:0016616', 0.008069923925358985),\n"," ('GO:0016709', 0.006775036418711125),\n"," ('GO:0032991', 0.006358822577288598),\n"," ('GO:0016758', 0.004832705158739335)]"]},"execution_count":18,"metadata":{},"output_type":"execute_result"}],"source":["# Degree Centrality - % of neighbors out of all nodes\n","GO_degree_centrality = nx.centrality.degree_centrality(go_graph)\n","\n","# Top 5 nodes with highest degree centrality\n","(sorted(GO_degree_centrality.items(), key=lambda item: item[1], reverse=True))[:5]"]},{"cell_type":"code","execution_count":null,"id":"b9838000-c5d6-4c66-b4ed-0424bc468125","metadata":{"id":"b9838000-c5d6-4c66-b4ed-0424bc468125"},"outputs":[],"source":["# EXTREMELY SLOW!!\n","# Betweenness centrality - the number of times a node lies on the shortest path between other nodes, meaning it acts as a bridge\n","# GO_betweenness_centrality = nx.centrality.betweenness_centrality(go_graph)\n","\n","# Top 5 nodes with highest betweenness centrality\n","# (sorted(GO_betweenness_centrality.items(), key=lambda item: item[1], reverse=True))[:5]"]},{"cell_type":"code","execution_count":null,"id":"3749803f-a3dc-4264-9334-5b031110000c","metadata":{"id":"3749803f-a3dc-4264-9334-5b031110000c","outputId":"ed01c190-9966-42d3-879d-b8bfd15e3b38"},"outputs":[{"data":{"text/plain":["[('GO:0008150', 0.13249641986781371),\n"," ('GO:0009987', 0.09177009024191234),\n"," ('GO:0050789', 0.05450615735840855),\n"," ('GO:0003674', 0.052467978206565916),\n"," ('GO:0065007', 0.05023079543531534),\n"," ('GO:0008152', 0.04454669778432706),\n"," ('GO:0003824', 0.0426376148572408),\n"," ('GO:0071704', 0.04236188137222803)]"]},"execution_count":20,"metadata":{},"output_type":"execute_result"}],"source":["# Closeness centrality - their ‘closeness’ to all other nodes in the network\n","# the higher the closeness centrality of a node, the closer it is located to the center of the network.\n","closeness_centrality = nx.centrality.closeness_centrality(go_graph)\n","\n","# Top 5 nodes with highest closeness centrality\n","(sorted(closeness_centrality.items(), key=lambda item: item[1], reverse=True))[:8]"]},{"cell_type":"code","execution_count":null,"id":"9412e9f7-a91d-456e-965f-00d7de2ea482","metadata":{"id":"9412e9f7-a91d-456e-965f-00d7de2ea482","outputId":"4ea6f162-0ea9-43b3-e10c-b9273b5b2bad"},"outputs":[{"data":{"text/plain":["[('GO:0008150', 0.9570036605845309),\n"," ('GO:0008152', 0.23355013653192808),\n"," ('GO:0009987', 0.13004250621401445),\n"," ('GO:0065007', 0.05374101319562961),\n"," ('GO:0032502', 0.05273139959773638)]"]},"execution_count":21,"metadata":{},"output_type":"execute_result"}],"source":["# Eigenvector Centrality - how connected a node is to other important nodes in the network\n","# A high eigenvector centrality means that the node is connected to other nodes who themselves have high eigenvector centralities.\n","GO_eigenvector_centrality = nx.centrality.eigenvector_centrality(nx.DiGraph(go_graph))\n","\n","# Top 5 nodes with highest eigenvector centrality\n","(sorted(GO_eigenvector_centrality.items(), key=lambda item: item[1], reverse=True))[:5]"]},{"cell_type":"code","execution_count":null,"id":"548eb42e-00b7-49cd-ae16-646edda4579c","metadata":{"id":"548eb42e-00b7-49cd-ae16-646edda4579c","outputId":"c89d6d6a-d894-451d-e1c7-24a159c126af"},"outputs":[{"data":{"text/plain":["0.04148416899611462"]},"execution_count":22,"metadata":{},"output_type":"execute_result"},{"data":{"text/plain":["0.02700774532392186"]},"execution_count":22,"metadata":{},"output_type":"execute_result"},{"data":{"text/plain":["0.0002592562235609651"]},"execution_count":22,"metadata":{},"output_type":"execute_result"}],"source":["# The clustering coefficient of a node is defined as the probability that two randomly selected friends of are friends with each other.\n","# As a result, the average clustering coefficient is the average of clustering coefficients of all the nodes.\n","nx.average_clustering(nx.DiGraph(BP_subgraph))\n","nx.average_clustering(nx.DiGraph(CC_subgraph))\n","nx.average_clustering(nx.DiGraph(MF_subgraph))"]},{"cell_type":"code","execution_count":null,"id":"5b6207ec-304b-4686-8d02-b99626936ad6","metadata":{"id":"5b6207ec-304b-4686-8d02-b99626936ad6","outputId":"7d319e77-a481-4ccb-8232-051ebf1de422"},"outputs":[{"name":"stdout","output_type":"stream","text":["Number of bridges: 12,605\n","Total number of edges: 84,805\n","% of bridges: 14.863510%\n","\n","Number of bridges: 2,753\n","Total number of edges: 64,536\n","% of bridges: 4.265836%\n","\n","Number of bridges: 1,484\n","Total number of edges: 6,495\n","% of bridges: 22.848345%\n","\n","Number of bridges: 8,374\n","Total number of edges: 13,714\n","% of bridges: 61.061689%\n","\n"]}],"source":["# Bridges: Deleting the edge would cause A and B to lie in two different components\n","def check_bridges(graph):\n","    if not nx.has_bridges(graph.to_undirected()):\n","        print(\"No bridges in this graph\")\n","        return\n","\n","    bridges = nx.bridges(graph.to_undirected())\n","    num_bridges = len(list(bridges))\n","    print(f\"Number of bridges: {num_bridges:,}\")\n","    print(f\"Total number of edges: {graph.number_of_edges():,}\")\n","    print(f\"% of bridges: {num_bridges / graph.number_of_edges():%}\")\n","    print()\n","\n","check_bridges(go_graph)\n","\n","check_bridges(BP_subgraph)\n","check_bridges(CC_subgraph)\n","check_bridges(MF_subgraph)"]},{"cell_type":"code","execution_count":null,"id":"7a25e13c-e7cb-4836-a9a5-5b9da8435f56","metadata":{"id":"7a25e13c-e7cb-4836-a9a5-5b9da8435f56","outputId":"cdd02c60-33fb-404d-88a5-49647be0af26"},"outputs":[{"data":{"text/plain":["-0.12418034755903223"]},"execution_count":24,"metadata":{},"output_type":"execute_result"},{"data":{"text/plain":["0.11234612296704903"]},"execution_count":24,"metadata":{},"output_type":"execute_result"},{"data":{"text/plain":["-0.02642072932299072"]},"execution_count":24,"metadata":{},"output_type":"execute_result"},{"data":{"text/plain":["-0.1843306292754108"]},"execution_count":24,"metadata":{},"output_type":"execute_result"}],"source":["# Assortativity\n","# the preference for a network’s nodes to attach to others that are \"similar\" in some way.\n","# \"Similar\" here means having the same degree\n","nx.degree_assortativity_coefficient(go_graph)\n","\n","nx.degree_assortativity_coefficient(BP_subgraph)\n","nx.degree_assortativity_coefficient(CC_subgraph)\n","nx.degree_assortativity_coefficient(MF_subgraph)"]},{"cell_type":"code","execution_count":null,"id":"fd4ac3c5-be0e-4f21-88a0-b900afe62ec7","metadata":{"id":"fd4ac3c5-be0e-4f21-88a0-b900afe62ec7","outputId":"7bd0511b-9960-48ee-c59e-3afcac3c21d6"},"outputs":[{"data":{"text/plain":["6590"]},"execution_count":25,"metadata":{},"output_type":"execute_result"},{"data":{"text/plain":["6.562670713201821"]},"execution_count":25,"metadata":{},"output_type":"execute_result"}],"source":["# Network community\n","# A community is a group of nodes, so that nodes inside the group are connected with many more edges than between groups.\n","communities = nx.community.label_propagation_communities(go_graph.to_undirected())\n","\n","len(communities) # number of communities\n","np.mean([len(comm) for comm in communities]) # average number of members in each community"]},{"cell_type":"markdown","id":"5ea1bf35-a16f-4205-a8fb-ec5c77d98c80","metadata":{"id":"5ea1bf35-a16f-4205-a8fb-ec5c77d98c80"},"source":["# Preprocessing"]},{"cell_type":"markdown","id":"da84af8a-df57-4d4d-843a-fcae432ba371","metadata":{"jp-MarkdownHeadingCollapsed":true,"id":"da84af8a-df57-4d4d-843a-fcae432ba371"},"source":["## Preprocess Protein Sequences"]},{"cell_type":"code","execution_count":null,"id":"ac257e84-eae5-449e-82f0-46261d2a4ae3","metadata":{"id":"ac257e84-eae5-449e-82f0-46261d2a4ae3"},"outputs":[],"source":["def create_sequence_dataframe_from_fasta(fasta_sequences: list[Bio.SeqRecord.SeqRecord]) -> pd.DataFrame:\n","    \"\"\"Create DataFrame for train/test sequences\"\"\"\n","\n","    seq_record_dict = {\n","        \"EntryID\": [],\n","        \"description\": [],\n","        \"sequence\": [],\n","    }\n","    for seq_record in fasta_sequences:\n","        seq_record_dict[\"EntryID\"].append(seq_record.id)\n","        seq_record_dict[\"description\"].append(seq_record.description)\n","        seq_record_dict[\"sequence\"].append(str(seq_record.seq))\n","\n","    return pd.DataFrame(seq_record_dict)\n","\n","def augment_features_from_fasta_description(sequence_dataframe: pd.DataFrame) -> pd.DataFrame:\n","    \"\"\"Extract features from the fasta header text\"\"\"\n","    sequence_dataframe = sequence_dataframe.copy()\n","\n","    # Extract information from fasta headers\n","    description_parts = sequence_dataframe.description.str.split(r\"\\|\", n=2, expand=True)\n","    extracted_groups = sequence_dataframe.description.str.extract(r\"([A-Z0-9]+)_([A-Z0-9]+) (.*?) OS=\")\n","\n","    # Create features\n","    sequence_dataframe = sequence_dataframe.assign(\n","        db=description_parts[0].str[-2:],\n","        description=description_parts[2],\n","\n","        entry_name_prefix=extracted_groups[0],\n","        entry_name_suffix=extracted_groups[1],\n","        protein_name=extracted_groups[2],\n","\n","        organism_name=train_sequences.description.str.extract(r\"OS=(.+) OX=\"),            # The source organism's name\n","        organism_id=train_sequences.description.str.extract(r\"OX=(.+?) \").astype(float),  # The source organism's id\n","        gene_name=train_sequences.description.str.extract(r\"GN=(.+?) \"),                  # Optional: Might be empty\n","        protein_existence=train_sequences.description.str.extract(r\"PE=(.+?) \").astype(float),\n","        sequence_version=train_sequences.description.str.extract(r\"SV=(.+?)$\").astype(float),\n","    )\n","\n","    # Drop unneeded column\n","    return sequence_dataframe.drop(columns=\"description\")"]},{"cell_type":"code","execution_count":null,"id":"a94067e0-8851-4b43-aa40-1ae22e478f5f","metadata":{"id":"a94067e0-8851-4b43-aa40-1ae22e478f5f","outputId":"49094091-865c-405b-fff9-0146a54906df"},"outputs":[{"data":{"text/plain":["(142246, 11)"]},"execution_count":174,"metadata":{},"output_type":"execute_result"},{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>EntryID</th>\n","      <th>sequence</th>\n","      <th>db</th>\n","      <th>entry_name_prefix</th>\n","      <th>entry_name_suffix</th>\n","      <th>protein_name</th>\n","      <th>organism_id</th>\n","      <th>gene_name</th>\n","      <th>protein_existence</th>\n","      <th>sequence_version</th>\n","      <th>taxonomyID</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>P20536</td>\n","      <td>MNSVTVSHAPYTITYHDDWEPVMSQLVEFYNEVASWLLRDETSPIP...</td>\n","      <td>sp</td>\n","      <td>UNG</td>\n","      <td>VACCC</td>\n","      <td>Uracil-DNA glycosylase</td>\n","      <td>10249.0</td>\n","      <td>UNG</td>\n","      <td>1.0</td>\n","      <td>1.0</td>\n","      <td>10249</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>O73864</td>\n","      <td>MTEYRNFLLLFITSLSVIYPCTGISWLGLTINGSSVGWNQTHHCKL...</td>\n","      <td>sp</td>\n","      <td>WNT11</td>\n","      <td>DANRE</td>\n","      <td>Protein Wnt-11</td>\n","      <td>7955.0</td>\n","      <td>wnt11</td>\n","      <td>2.0</td>\n","      <td>1.0</td>\n","      <td>7955</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>O95231</td>\n","      <td>MRLSSSPPRGPQQLSSFGSVDWLSQSSCSGPTHTPRPADFSLGSLP...</td>\n","      <td>sp</td>\n","      <td>VENTX</td>\n","      <td>HUMAN</td>\n","      <td>Homeobox protein VENTX</td>\n","      <td>9606.0</td>\n","      <td>VENTX</td>\n","      <td>1.0</td>\n","      <td>1.0</td>\n","      <td>9606</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>A0A0B4J1F4</td>\n","      <td>MGGEAGADGPRGRVKSLGLVFEDESKGCYSSGETVAGHVLLEAAEP...</td>\n","      <td>sp</td>\n","      <td>ARRD4</td>\n","      <td>MOUSE</td>\n","      <td>Arrestin domain-containing protein 4</td>\n","      <td>10090.0</td>\n","      <td>Arrdc4</td>\n","      <td>1.0</td>\n","      <td>1.0</td>\n","      <td>10090</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>P54366</td>\n","      <td>MVETNSPPAGYTLKRSPSDLGEQQQPPRQISRSPGNTAAYHLTTAM...</td>\n","      <td>sp</td>\n","      <td>GSC</td>\n","      <td>DROME</td>\n","      <td>Homeobox protein goosecoid</td>\n","      <td>7227.0</td>\n","      <td>Gsc</td>\n","      <td>2.0</td>\n","      <td>2.0</td>\n","      <td>7227</td>\n","    </tr>\n","  </tbody>\n","</table>\n","</div>"],"text/plain":["      EntryID                                           sequence  db  \\\n","0      P20536  MNSVTVSHAPYTITYHDDWEPVMSQLVEFYNEVASWLLRDETSPIP...  sp   \n","1      O73864  MTEYRNFLLLFITSLSVIYPCTGISWLGLTINGSSVGWNQTHHCKL...  sp   \n","2      O95231  MRLSSSPPRGPQQLSSFGSVDWLSQSSCSGPTHTPRPADFSLGSLP...  sp   \n","3  A0A0B4J1F4  MGGEAGADGPRGRVKSLGLVFEDESKGCYSSGETVAGHVLLEAAEP...  sp   \n","4      P54366  MVETNSPPAGYTLKRSPSDLGEQQQPPRQISRSPGNTAAYHLTTAM...  sp   \n","\n","  entry_name_prefix entry_name_suffix                          protein_name  \\\n","0               UNG             VACCC                Uracil-DNA glycosylase   \n","1             WNT11             DANRE                        Protein Wnt-11   \n","2             VENTX             HUMAN                Homeobox protein VENTX   \n","3             ARRD4             MOUSE  Arrestin domain-containing protein 4   \n","4               GSC             DROME            Homeobox protein goosecoid   \n","\n","   organism_id gene_name  protein_existence  sequence_version  taxonomyID  \n","0      10249.0       UNG                1.0               1.0       10249  \n","1       7955.0     wnt11                2.0               1.0        7955  \n","2       9606.0     VENTX                1.0               1.0        9606  \n","3      10090.0    Arrdc4                1.0               1.0       10090  \n","4       7227.0       Gsc                2.0               2.0        7227  "]},"execution_count":174,"metadata":{},"output_type":"execute_result"}],"source":["# Preprocess the train sequences\n","train_sequences = create_sequence_dataframe_from_fasta(train_sequences_fasta)\n","train_sequences = augment_features_from_fasta_description(train_sequences)\n","train_sequences = train_sequences.merge(train_taxonomy, on=\"EntryID\", how=\"left\")\n","\n","train_sequences.shape\n","train_sequences.head()"]},{"cell_type":"code","execution_count":null,"id":"2c1b32b2-6067-4cbd-9eab-d15da78692a7","metadata":{"id":"2c1b32b2-6067-4cbd-9eab-d15da78692a7"},"outputs":[],"source":["# Unfortunately, we have to discard most of the features because the test data does not have those information in its fasta header\n","train_sequences = train_sequences[[\"EntryID\", \"sequence\", \"organism_id\", \"taxonomyID\"]].copy()"]},{"cell_type":"code","execution_count":null,"id":"08940c0b-957c-49be-b370-5840d632f7e1","metadata":{"id":"08940c0b-957c-49be-b370-5840d632f7e1","outputId":"86961fc6-87a1-48a6-f94c-de2186e73145"},"outputs":[{"data":{"text/plain":["(141865, 4)"]},"execution_count":206,"metadata":{},"output_type":"execute_result"},{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>EntryID</th>\n","      <th>sequence</th>\n","      <th>organism_id</th>\n","      <th>taxonomyID</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>Q9CQV8</td>\n","      <td>MTMDKSELVQKAKLAEQAERYDDMAAAMKAVTEQGHELSNEERNLL...</td>\n","      <td>10090</td>\n","      <td>10090.0</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>P62259</td>\n","      <td>MDDREDLVYQAKLAEQAERYDEMVESMKKVAGMDVELTVEERNLLS...</td>\n","      <td>10090</td>\n","      <td>10090.0</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>P68510</td>\n","      <td>MGDREQLLQRARLAEQAERYDDMASAMKAVTELNEPLSNEDRNLLS...</td>\n","      <td>10090</td>\n","      <td>10090.0</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>P61982</td>\n","      <td>MVDREQLVQKARLAEQAERYDDMAAAMKNVTELNEPLSNEERNLLS...</td>\n","      <td>10090</td>\n","      <td>10090.0</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>O70456</td>\n","      <td>MERASLIQKAKLAEQAERYEDMAAFMKSAVEKGEELSCEERNLLSV...</td>\n","      <td>10090</td>\n","      <td>10090.0</td>\n","    </tr>\n","  </tbody>\n","</table>\n","</div>"],"text/plain":["  EntryID                                           sequence organism_id  \\\n","0  Q9CQV8  MTMDKSELVQKAKLAEQAERYDDMAAAMKAVTEQGHELSNEERNLL...       10090   \n","1  P62259  MDDREDLVYQAKLAEQAERYDEMVESMKKVAGMDVELTVEERNLLS...       10090   \n","2  P68510  MGDREQLLQRARLAEQAERYDDMASAMKAVTELNEPLSNEDRNLLS...       10090   \n","3  P61982  MVDREQLVQKARLAEQAERYDDMAAAMKNVTELNEPLSNEERNLLS...       10090   \n","4  O70456  MERASLIQKAKLAEQAERYEDMAAFMKSAVEKGEELSCEERNLLSV...       10090   \n","\n","   taxonomyID  \n","0     10090.0  \n","1     10090.0  \n","2     10090.0  \n","3     10090.0  \n","4     10090.0  "]},"execution_count":206,"metadata":{},"output_type":"execute_result"}],"source":["# Preprocess the test sequences\n","test_sequences = create_sequence_dataframe_from_fasta(test_sequences_fasta)\n","test_sequences = (test_sequences\n","    .assign(organism_id=test_sequences.description.str.split(\"\\t\", expand=True)[1])\n","    .drop(columns=\"description\")\n",")\n","test_sequences = test_sequences.merge(train_taxonomy, on=\"EntryID\", how=\"left\")\n","\n","test_sequences.shape\n","test_sequences.head()"]},{"cell_type":"code","execution_count":null,"id":"6c7fd670-5aa6-49c8-80bb-b5d32d933a06","metadata":{"id":"6c7fd670-5aa6-49c8-80bb-b5d32d933a06"},"outputs":[],"source":["# Note: The train_sequence and train_terms contains the same set of IDs!\n","# set(train_sequences.EntryID) == set(train_terms.EntryID) # True"]},{"cell_type":"code","execution_count":null,"id":"d7c1f044-c8b0-4b7a-b95f-ee367c813a4c","metadata":{"id":"d7c1f044-c8b0-4b7a-b95f-ee367c813a4c","outputId":"a425ab19-43e8-4c76-a4eb-b65ef1cc1534"},"outputs":[{"data":{"text/plain":["(5363863, 3)"]},"execution_count":187,"metadata":{},"output_type":"execute_result"},{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>EntryID</th>\n","      <th>term</th>\n","      <th>aspect</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0008152</td>\n","      <td>BPO</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0034655</td>\n","      <td>BPO</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0072523</td>\n","      <td>BPO</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0044270</td>\n","      <td>BPO</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0006753</td>\n","      <td>BPO</td>\n","    </tr>\n","  </tbody>\n","</table>\n","</div>"],"text/plain":["      EntryID        term aspect\n","0  A0A009IHW8  GO:0008152    BPO\n","1  A0A009IHW8  GO:0034655    BPO\n","2  A0A009IHW8  GO:0072523    BPO\n","3  A0A009IHW8  GO:0044270    BPO\n","4  A0A009IHW8  GO:0006753    BPO"]},"execution_count":187,"metadata":{},"output_type":"execute_result"}],"source":["# Labels of the train sequences\n","train_terms.shape\n","train_terms.head()"]},{"cell_type":"markdown","id":"dd684aa9-485f-4764-a6d9-3409569cf9bf","metadata":{"id":"dd684aa9-485f-4764-a6d9-3409569cf9bf"},"source":["## Preprocess GO terms"]},{"cell_type":"code","execution_count":null,"id":"d70a5ced-9b1b-4729-8881-2dedf0bf0ce1","metadata":{"id":"d70a5ced-9b1b-4729-8881-2dedf0bf0ce1","outputId":"15f34dd9-ad71-4c84-9135-7261979956bf"},"outputs":[{"data":{"text/plain":["torch.Size([2, 10, 1024])"]},"execution_count":220,"metadata":{},"output_type":"execute_result"}],"source":["# Encode texts (example)\n","inputs = biogpt_tokenizer([\"Hello, my dog is cute\", \"Algorithms in Structural Bioinformatics\"], return_tensors=\"pt\", padding=True)\n","outputs = biogpt_model(**inputs)\n","\n","last_hidden_states = outputs.last_hidden_state\n","last_hidden_states.shape"]},{"cell_type":"code","execution_count":null,"id":"819feb2d-1999-4d3d-8965-06a5b7ee7a64","metadata":{"id":"819feb2d-1999-4d3d-8965-06a5b7ee7a64","outputId":"8cf3dcf0-bd2f-46ec-c2db-a0c979ac0626"},"outputs":[{"data":{"text/plain":["<networkx.classes.multidigraph.MultiDiGraph at 0x16d25f040>"]},"execution_count":209,"metadata":{},"output_type":"execute_result"}],"source":["go_graph."]},{"cell_type":"code","execution_count":null,"id":"ecf5a4df-d33f-4bbb-a22a-1100b408ed27","metadata":{"id":"ecf5a4df-d33f-4bbb-a22a-1100b408ed27"},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"id":"cb8517c9-a150-4122-9a09-9ea67a340056","metadata":{"id":"cb8517c9-a150-4122-9a09-9ea67a340056"},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"id":"f37531df-2dce-42cd-9a40-b858175b7b67","metadata":{"id":"f37531df-2dce-42cd-9a40-b858175b7b67"},"outputs":[],"source":[]},{"cell_type":"markdown","id":"4e5c24dd-25e1-4102-8fd5-9721908d343b","metadata":{"id":"4e5c24dd-25e1-4102-8fd5-9721908d343b"},"source":["# Feature Extraction Components"]},{"cell_type":"markdown","id":"14db4c70-44e7-46d5-9f56-1730ff612db8","metadata":{"id":"14db4c70-44e7-46d5-9f56-1730ff612db8"},"source":["### Naive\n","The score that a protein $P_j$ is associated with a GO term $G_i$ is defined as $G_i$'s relative frequency in the training data."]},{"cell_type":"code","execution_count":null,"id":"ed29a4a8-07f1-434e-bf0a-51b7a1491eb0","metadata":{"id":"ed29a4a8-07f1-434e-bf0a-51b7a1491eb0","outputId":"01cdbb34-ebcc-4db3-ea58-93eab409a2ce"},"outputs":[{"data":{"text/plain":["term\n","GO:0005575    1.732184e-02\n","GO:0008150    1.719097e-02\n","GO:0110165    1.701870e-02\n","GO:0003674    1.466052e-02\n","GO:0005622    1.319665e-02\n","                  ...     \n","GO:0031772    1.864328e-07\n","GO:0042324    1.864328e-07\n","GO:0031771    1.864328e-07\n","GO:0051041    1.864328e-07\n","GO:0102628    1.864328e-07\n","Name: proportion, Length: 31466, dtype: float64"]},"execution_count":27,"metadata":{},"output_type":"execute_result"}],"source":["# Naive\n","naive_component = train_terms[\"term\"].value_counts(dropna=False, normalize=True)\n","naive_component"]},{"cell_type":"code","execution_count":null,"id":"33c9a00b-73a3-42d7-8d17-f4c76f18134a","metadata":{"id":"33c9a00b-73a3-42d7-8d17-f4c76f18134a"},"outputs":[],"source":[]},{"cell_type":"markdown","id":"800849f5-fa36-43b3-a6ab-3477f1ada490","metadata":{"id":"800849f5-fa36-43b3-a6ab-3477f1ada490"},"source":["# Baseline Model"]},{"cell_type":"code","execution_count":null,"id":"16aadbd5-2133-4dcd-b109-7923bc0e55e7","metadata":{"id":"16aadbd5-2133-4dcd-b109-7923bc0e55e7","outputId":"80f80d55-8124-48da-a67f-ea9cb7823ee4"},"outputs":[{"name":"stdout","output_type":"stream","text":["device(type='cpu')\n"]}],"source":["# Setup device\n","DEVICE = torch.device('cuda' if torch.cuda.is_available() else 'cpu')\n","display(DEVICE)\n","\n","# Fix random seed for reproducibility\n","SEED = 42\n","def set_seed(seed: int=SEED) -> None:\n","    random.seed(seed)\n","    np.random.seed(seed)\n","    torch.manual_seed(seed)\n","    if torch.cuda.is_available():\n","        torch.cuda.manual_seed(seed)\n","        torch.cuda.manual_seed_all(seed)\n","    torch.backends.cudnn.benchmark = False\n","    torch.backends.cudnn.deterministic = True\n","    return None\n","\n","set_seed()"]},{"cell_type":"code","execution_count":null,"id":"e5ba5450-bad7-4c64-b63b-9aae2f1c5d53","metadata":{"id":"e5ba5450-bad7-4c64-b63b-9aae2f1c5d53","outputId":"ac1a571f-8f93-4240-e3d7-1cffd6439ec5"},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>EntryID</th>\n","      <th>sequence</th>\n","      <th>organism_id</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>P20536</td>\n","      <td>MNSVTVSHAPYTITYHDDWEPVMSQLVEFYNEVASWLLRDETSPIP...</td>\n","      <td>10249.0</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>O73864</td>\n","      <td>MTEYRNFLLLFITSLSVIYPCTGISWLGLTINGSSVGWNQTHHCKL...</td>\n","      <td>7955.0</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>O95231</td>\n","      <td>MRLSSSPPRGPQQLSSFGSVDWLSQSSCSGPTHTPRPADFSLGSLP...</td>\n","      <td>9606.0</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>A0A0B4J1F4</td>\n","      <td>MGGEAGADGPRGRVKSLGLVFEDESKGCYSSGETVAGHVLLEAAEP...</td>\n","      <td>10090.0</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>P54366</td>\n","      <td>MVETNSPPAGYTLKRSPSDLGEQQQPPRQISRSPGNTAAYHLTTAM...</td>\n","      <td>7227.0</td>\n","    </tr>\n","    <tr>\n","      <th>...</th>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","    </tr>\n","    <tr>\n","      <th>142241</th>\n","      <td>A0A286YAI0</td>\n","      <td>METEVDDFPGKASIFSQVNPLYSNNMKLCEAERYDFQHSEPKTMKS...</td>\n","      <td>7955.0</td>\n","    </tr>\n","    <tr>\n","      <th>142242</th>\n","      <td>A0A1D5NUC4</td>\n","      <td>MSAAASAEMIETPPVLNFEEIDYKEIEVEEVVGRGAFGVVCKAKWR...</td>\n","      <td>9031.0</td>\n","    </tr>\n","    <tr>\n","      <th>142243</th>\n","      <td>Q5RGB0</td>\n","      <td>MADKGPILTSVIIFYLSIGAAIFQILEEPNLNSAVDDYKNKTNNLL...</td>\n","      <td>7955.0</td>\n","    </tr>\n","    <tr>\n","      <th>142244</th>\n","      <td>A0A2R8QMZ5</td>\n","      <td>MGRKKIQITRIMDERNRQVTFTKRKFGLMKKAYELSVLCDCEIALI...</td>\n","      <td>7955.0</td>\n","    </tr>\n","    <tr>\n","      <th>142245</th>\n","      <td>A0A8I6GHU0</td>\n","      <td>HCISSLKLTAFFKRSFLLSPEKHLVLLRDGRTLIGFLRSIDQFANL...</td>\n","      <td>10116.0</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>142246 rows × 3 columns</p>\n","</div>"],"text/plain":["           EntryID                                           sequence  \\\n","0           P20536  MNSVTVSHAPYTITYHDDWEPVMSQLVEFYNEVASWLLRDETSPIP...   \n","1           O73864  MTEYRNFLLLFITSLSVIYPCTGISWLGLTINGSSVGWNQTHHCKL...   \n","2           O95231  MRLSSSPPRGPQQLSSFGSVDWLSQSSCSGPTHTPRPADFSLGSLP...   \n","3       A0A0B4J1F4  MGGEAGADGPRGRVKSLGLVFEDESKGCYSSGETVAGHVLLEAAEP...   \n","4           P54366  MVETNSPPAGYTLKRSPSDLGEQQQPPRQISRSPGNTAAYHLTTAM...   \n","...            ...                                                ...   \n","142241  A0A286YAI0  METEVDDFPGKASIFSQVNPLYSNNMKLCEAERYDFQHSEPKTMKS...   \n","142242  A0A1D5NUC4  MSAAASAEMIETPPVLNFEEIDYKEIEVEEVVGRGAFGVVCKAKWR...   \n","142243      Q5RGB0  MADKGPILTSVIIFYLSIGAAIFQILEEPNLNSAVDDYKNKTNNLL...   \n","142244  A0A2R8QMZ5  MGRKKIQITRIMDERNRQVTFTKRKFGLMKKAYELSVLCDCEIALI...   \n","142245  A0A8I6GHU0  HCISSLKLTAFFKRSFLLSPEKHLVLLRDGRTLIGFLRSIDQFANL...   \n","\n","        organism_id  \n","0           10249.0  \n","1            7955.0  \n","2            9606.0  \n","3           10090.0  \n","4            7227.0  \n","...             ...  \n","142241       7955.0  \n","142242       9031.0  \n","142243       7955.0  \n","142244       7955.0  \n","142245      10116.0  \n","\n","[142246 rows x 3 columns]"]},"execution_count":222,"metadata":{},"output_type":"execute_result"}],"source":["train_sequences.sequence"]},{"cell_type":"code","execution_count":null,"id":"7404a424-b17e-4ec0-adb2-fc40fabc1550","metadata":{"id":"7404a424-b17e-4ec0-adb2-fc40fabc1550"},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"id":"e6a5d3cc-51b9-44d8-a3bb-9ca6ced8b549","metadata":{"id":"e6a5d3cc-51b9-44d8-a3bb-9ca6ced8b549"},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"id":"9c22ba95-ba6a-42e7-8824-7229ff8620df","metadata":{"id":"9c22ba95-ba6a-42e7-8824-7229ff8620df"},"outputs":[],"source":[]}],"metadata":{"kernelspec":{"display_name":"Python [conda env:cafa-5]","language":"python","name":"conda-env-cafa-5-py"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.10.12"},"colab":{"provenance":[]}},"nbformat":4,"nbformat_minor":5}